

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=&#34;auto&#34;>



<head>
  <meta charset="UTF-8">
  <link rel="apple-touch-icon" sizes="76x76" href="/img/favicon.png">
  <link rel="icon" href="/img/favicon.png">
  <meta name="viewport"
        content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="description" content="学习snakemake 官方文档">
  <meta name="author" content="wutao">
  <meta name="keywords" content="">
  
  <title>使用Snakemake搭建流程 - Hexo</title>

  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css" />


  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/github-markdown-css@4.0.0/github-markdown.min.css" />
  <link  rel="stylesheet" href="/lib/hint/hint.min.css" />

  
    
    
      
      <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/highlight.js@10.6.0/styles/github-gist.min.css" />
    
  

  
    <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.5.7/dist/jquery.fancybox.min.css" />
  



<!-- 主题依赖的图标库，不要自行修改 -->

<link rel="stylesheet" href="//at.alicdn.com/t/font_1749284_ba1fz6golrf.css">



<link rel="stylesheet" href="//at.alicdn.com/t/font_1736178_kmeydafke9r.css">


<link  rel="stylesheet" href="/css/main.css" />

<!-- 自定义样式保持在最底部 -->


  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    var CONFIG = {"hostname":"example.com","root":"/","version":"1.8.10","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"right","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"copy_btn":true,"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"baidu":null,"google":null,"gtag":null,"tencent":{"sid":null,"cid":null},"woyaola":null,"cnzz":null,"leancloud":{"app_id":null,"app_key":null,"server_url":null}}};
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
<meta name="generator" content="Hexo 5.4.0"></head>


<body>
  <header style="height: 70vh;">
    <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand"
       href="/">&nbsp;<strong>wutao's blog</strong>&nbsp;</a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/">
                <i class="iconfont icon-home-fill"></i>
                首页
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/">
                <i class="iconfont icon-archive-fill"></i>
                归档
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/">
                <i class="iconfont icon-category-fill"></i>
                分类
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/">
                <i class="iconfont icon-tags-fill"></i>
                标签
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/">
                <i class="iconfont icon-user-fill"></i>
                关于
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" data-toggle="modal" data-target="#modalSearch">&nbsp;<i
                class="iconfont icon-search"></i>&nbsp;</a>
          </li>
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self">&nbsp;<i
                class="iconfont icon-dark" id="color-toggle-icon"></i>&nbsp;</a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

    <div class="banner" id="banner" parallax=true
         style="background: url('/img/default.png') no-repeat center center;
           background-size: cover;">
      <div class="full-bg-img">
        <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
          <div class="page-header text-center fade-in-up">
            <span class="h2" id="subtitle" title="使用Snakemake搭建流程">
              
            </span>

            
              <div class="mt-3">
  
    <span class="post-meta mr-2">
      <i class="iconfont icon-author" aria-hidden="true"></i>
      wutao
    </span>
  
  
    <span class="post-meta">
      <i class="iconfont icon-date-fill" aria-hidden="true"></i>
      <time datetime="2021-03-19 10:00" pubdate>
        2021年3月19日 上午
      </time>
    </span>
  
</div>

<div class="mt-1">
  
    
    <span class="post-meta mr-2">
      <i class="iconfont icon-chart"></i>
      4.1k 字
    </span>
  

  
    
    <span class="post-meta mr-2">
      <i class="iconfont icon-clock-fill"></i>
      
      
      59
       分钟
    </span>
  

  
  
</div>

            
          </div>

          
        </div>
      </div>
    </div>
  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="d-none d-lg-block col-lg-2"></div>
    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div class="py-5" id="board">
          <article class="post-content mx-auto">
            <!-- SEO header -->
            <h1 style="display: none">使用Snakemake搭建流程</h1>
            
              <p class="note note-info">
                
                  本文最后更新于：2 个月前
                
              </p>
            
            <div class="markdown-body">
              <p>学习<code>snakemake</code> 官方文档</p>
<span id="more"></span>

<p>创建一个<code>snakemake_tutorial</code>目录,并下载示例数据：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs bash">wget https://github.com/snakemake/snakemake-tutorial-data/archive/v5.24.1.tar.gz<br><br>tar --wildcards -xf snakemake-tutorial-data-5.24.1.tar.gz --strip 1 <span class="hljs-string">&quot;*/data&quot;</span> <span class="hljs-string">&quot;*/environment.yaml&quot;</span><br><span class="hljs-comment">##--wildcards 根据通配符来提取压缩文件,这里是只提取data文件夹和environment.yaml文件</span><br></code></pre></td></tr></table></figure>

<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><code class="hljs stylus">(snakemake-tutorial) -bash-<span class="hljs-number">4.2</span>$ tree .<br>.<br>├── data<br>│   ├── genome<span class="hljs-selector-class">.fa</span><br>│   ├── genome<span class="hljs-selector-class">.fa</span><span class="hljs-selector-class">.amb</span><br>│   ├── genome<span class="hljs-selector-class">.fa</span><span class="hljs-selector-class">.ann</span><br>│   ├── genome<span class="hljs-selector-class">.fa</span><span class="hljs-selector-class">.bwt</span><br>│   ├── genome<span class="hljs-selector-class">.fa</span><span class="hljs-selector-class">.fai</span><br>│   ├── genome<span class="hljs-selector-class">.fa</span><span class="hljs-selector-class">.pac</span><br>│   ├── genome<span class="hljs-selector-class">.fa</span><span class="hljs-selector-class">.sa</span><br>│   └── samples<br>│       ├── A<span class="hljs-selector-class">.fastq</span><br>│       ├── B<span class="hljs-selector-class">.fastq</span><br>│       └── C<span class="hljs-selector-class">.fastq</span><br>├── environment<span class="hljs-selector-class">.yaml</span><br>└── snakemake-tutorial-data-<span class="hljs-number">5.24</span>.<span class="hljs-number">1</span><span class="hljs-selector-class">.tar</span><span class="hljs-selector-class">.gz</span><br><br><span class="hljs-number">2</span> directories, <span class="hljs-number">12</span> files<br></code></pre></td></tr></table></figure>

<p>其中<code>environment.yaml</code>文件是用来创建所需的<code>conda</code>环境:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs bash">cat environment.yaml<br><br>channels:<br>  - bioconda<br>  - conda-forge<br>dependencies:<br>  - snakemake-minimal &gt;=5.24.1<br>  - jinja2 =2.11<br>  - networkx =2.5<br>  - matplotlib =3.3<br>  - graphviz =2.42<br>  - bcftools =1.9<br>  - samtools =1.9<br>  - bwa =0.7<br>  - pysam =0.15<br></code></pre></td></tr></table></figure>

<p>使用该配置文件创建<code>snakemake-tutorial</code>的环境(使用mamba代替conda来加速下载)：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">mamba env create --name snakemake-tutorial --file environment.yaml<br><br>conda activate snakemake-tutorial<br></code></pre></td></tr></table></figure>

<h2 id="基础：以一个生物信息学流程为例"><a href="#基础：以一个生物信息学流程为例" class="headerlink" title="基础：以一个生物信息学流程为例"></a>基础：以一个生物信息学流程为例</h2><p>一个<code>Snakemake</code>流程由<code>Snakefile</code>文件中的一系列规则(rules)来创建；这些规则通过说明<strong>如何从输入文件得到输出文件</strong>来将流程分解成多个小的步骤,<code>Snakemake</code>会通过匹配文件名来自动的决定规则间的依赖关系</p>
<p>接下来以一个生物信息学的流程为例来学习<code>Snakemake</code>流程的搭建</p>
<p>这个流程做的工作为：将测序的reads匹配到参考基因组上,并且检测匹配上的reads的变异</p>
<h3 id="第一步：Mapping-reads"><a href="#第一步：Mapping-reads" class="headerlink" title="第一步：Mapping reads"></a>第一步：Mapping reads</h3><p>第一个<code>Snakemake</code>规则将给定样本的测序reads回帖到给定的参考基因组上去,使用的工具为<a target="_blank" rel="noopener" href="http://bio-bwa.sourceforge.net/">bwa的mem算法</a></p>
<p>创建一个<code>Snakefile</code>文件,写上下面的规则：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs python">rule bwa_map:<br>    <span class="hljs-built_in">input</span>:<br>        <span class="hljs-string">&quot;data/genome.fa&quot;</span>,<br>        <span class="hljs-string">&quot;data/samples/A.fastq&quot;</span><br>    output:<br>        <span class="hljs-string">&quot;mapped_reads/A.bam&quot;</span><br>    shell:<br>        <span class="hljs-string">&quot;bwa mem &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<p>一个<code>Snakemake</code>规则有一个名字,这里是<code>bwa_map</code>;还有一些指令,上面的例子里是<code>input</code>, <code>output</code>和<code>shell</code>;<code>input</code>和<code>output</code>指令中是一系列的文件名(python 字符串),指定了输入和输出文件(如果有多个文件,用逗号分割);<code>shell</code>指令也是一个字符串,表示需要执行的shell命令,在shell命令字符串中可以使用花括号来指代规则中的其他部分,比如这里使用<code>&#123;input&#125;</code>来指代<code>input</code>指令中的内容,使用<code>&#123;output&#125;</code>指代<code>output</code>指令中的内容;上面的<code>input</code>里面有两个字符串,这时<code>snakemake</code>替代<code>&#123;input&#125;</code>时会用空格分隔开两个输入文件 </p>
<p>接下来可以执行这个流程：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake --cores 1 <br><br>Building DAG of <span class="hljs-built_in">jobs</span>...<br>Using shell: /usr/bin/bash<br>Provided cores: 1 (use --cores to define parallelism)<br>Rules claiming more threads will be scaled down.<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        1       bwa_map<br>        1<br>Select <span class="hljs-built_in">jobs</span> to execute...<br><br>[Sat Mar 20 18:17:43 2021]<br>rule bwa_map:<br>    input: data/genome.fa, data/samples/A.fastq<br>    output: mapped_reads/A.bam<br>    jobid: 0<br><br>[M::bwa_idx_load_from_disk] <span class="hljs-built_in">read</span> 0 ALT contigs<br>[M::process] <span class="hljs-built_in">read</span> 25000 sequences (2525000 bp)...<br>[M::mem_process_seqs] Processed 25000 reads <span class="hljs-keyword">in</span> 1.267 CPU sec, 1.267 real sec<br>[main] Version: 0.7.17-r1188<br>[main] CMD: bwa mem data/genome.fa data/samples/A.fastq<br>[main] Real time: 1.757 sec; CPU: 1.318 sec<br>[Sat Mar 20 18:17:44 2021]<br>Finished job 0.<br>1 of 1 steps (100%) <span class="hljs-keyword">done</span><br>Complete <span class="hljs-built_in">log</span>: /slst/home/wutao2/snakemake_tutorial/.snakemake/<span class="hljs-built_in">log</span>/2021-03-20T181738.569944.snakemake.log<br><br>tree .<br>.<br>├── data<br>│   ├── genome.fa<br>│   ├── genome.fa.amb<br>│   ├── genome.fa.ann<br>│   ├── genome.fa.bwt<br>│   ├── genome.fa.fai<br>│   ├── genome.fa.pac<br>│   ├── genome.fa.sa<br>│   └── samples<br>│       ├── A.fastq<br>│       ├── B.fastq<br>│       └── C.fastq<br>├── environment.yaml<br>├── mapped_reads<br>│   └── A.bam<br>├── Snakefile<br>└── snakemake-tutorial-data-5.24.1.tar.gz<br></code></pre></td></tr></table></figure>

<p>也可以使用<code>-n</code>或者<code>--dry-run</code>参数使snakemake显示执行的”计划“(没有真正的执行流程);使用<code>-p</code>参数来打印需要执行的命令：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake -np<br><br>Building DAG of <span class="hljs-built_in">jobs</span>...<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        1       bwa_map<br>        1<br><br>[Sat Mar 20 18:25:11 2021]<br>rule bwa_map:<br>    input: data/genome.fa, data/samples/A.fastq<br>    output: mapped_reads/A.bam<br>    jobid: 0<br><br>bwa mem data/genome.fa data/samples/A.fastq | samtools view -Sb - &gt; mapped_reads/A.bam<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        1       bwa_map<br>        1<br>This was a dry-run (flag -n). The order of <span class="hljs-built_in">jobs</span> does not reflect the order of execution.<br></code></pre></td></tr></table></figure>

<h3 id="第二步：使规则适用的范围更广"><a href="#第二步：使规则适用的范围更广" class="headerlink" title="第二步：使规则适用的范围更广"></a>第二步：使规则适用的范围更广</h3><p>上面的规则只能对单个样本<code>data/samples/A.fastq</code>适用,在snakemake中可以使用通配符(wildcard)来扩展规则的适用范围：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs bash">rule bwa_map:<br>    input:<br>        <span class="hljs-string">&quot;data/genome.fa&quot;</span>,<br>        <span class="hljs-string">&quot;data/samples/&#123;sample&#125;.fastq&quot;</span><br>    output:<br>        <span class="hljs-string">&quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br>    shell:<br>        <span class="hljs-string">&quot;bwa mem &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<p><code>Snakemake</code>会将<code>output</code>中的<code>&#123;sample&#125;</code>替换成一个合适的值,并且将<code>input</code>中的<code>&#123;sample&#125;</code>也替换成同样的值,我们在运行流程就需要指定输出文件的名称(这样snakemake才知道如何替换通配符)：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake -np mapped_reads/B.bam<br><br>Building DAG of <span class="hljs-built_in">jobs</span>...<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        1       bwa_map<br>        1<br><br>[Sat Mar 20 18:50:12 2021]<br>rule bwa_map:<br>    input: data/genome.fa, data/samples/B.fastq<br>    output: mapped_reads/B.bam<br>    jobid: 0<br>    wildcards: sample=B<br><br>bwa mem data/genome.fa data/samples/B.fastq | samtools view -Sb - &gt; mapped_reads/B.bam<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        1       bwa_map<br>        1<br>This was a dry-run (flag -n). The order of <span class="hljs-built_in">jobs</span> does not reflect the order of execution.<br></code></pre></td></tr></table></figure>
<p>这个时候snakemake就将<code>&#123;sample&#125;</code>替换成<code>B</code>了<br>也可以同时生成多个文件：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake -np mapped_reads/A.bam mapped_reads/B.bam<br><span class="hljs-comment">##或snakemake -np mapped_reads/&#123;A,B&#125;.bam</span><br><br>Building DAG of <span class="hljs-built_in">jobs</span>...<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        2       bwa_map<br>        2<br><br>[Sat Mar 20 18:52:00 2021]<br>rule bwa_map:<br>    input: data/genome.fa, data/samples/B.fastq<br>    output: mapped_reads/B.bam<br>    jobid: 0<br>    wildcards: sample=B<br><br>bwa mem data/genome.fa data/samples/B.fastq | samtools view -Sb - &gt; mapped_reads/B.bam<br><br>[Sat Mar 20 18:52:00 2021]<br>rule bwa_map:<br>    input: data/genome.fa, data/samples/A.fastq<br>    output: mapped_reads/A.bam<br>    jobid: 1<br>    wildcards: sample=A<br><br>bwa mem data/genome.fa data/samples/A.fastq | samtools view -Sb - &gt; mapped_reads/A.bam<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        2       bwa_map<br>        2<br>This was a dry-run (flag -n). The order of <span class="hljs-built_in">jobs</span> does not reflect the order of execution.<br></code></pre></td></tr></table></figure>

<h3 id="第三步：Sorting-read-alignments"><a href="#第三步：Sorting-read-alignments" class="headerlink" title="第三步：Sorting read alignments"></a>第三步：Sorting read alignments</h3><p>接下来需要使用<code>samtools</code>中的<code>sort</code>命令来对BAM文件进行排序,将下面的规则写到刚才的<code>bwa_map</code>规则的下面：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs bash">rule samtools_sort:<br>    input:<br>        <span class="hljs-string">&quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br>    output:<br>        <span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;_sorted.bam&quot;</span><br>    shell:<br>        <span class="hljs-string">&quot;samtools sort -T sorted_reads/&#123;wildcards.sample&#125; -O bam &#123;input&#125; &gt; &#123;output&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<p>这个规则的输入文件是刚才<code>bwa_map</code>规则的输出文件;这里面需要注意的是在shell命令中可以通过<code>wildcards</code>对象来获取不同通配符的值(即wildcards对象的属性)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake -np sorted_reads/B_sorted.bam<br>Building DAG of <span class="hljs-built_in">jobs</span>...<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        1       bwa_map<br>        1       samtools_sort<br>        2<br><br>[Sun Mar 21 16:06:07 2021]<br>rule bwa_map:<br>    input: data/genome.fa, data/samples/B.fastq<br>    output: mapped_reads/B.bam<br>    jobid: 1<br>    wildcards: sample=B<br><br>bwa mem data/genome.fa data/samples/B.fastq | samtools view -Sb - &gt; mapped_reads/B.bam<br><br>[Sun Mar 21 16:06:07 2021]<br>rule samtools_sort:<br>    input: mapped_reads/B.bam<br>    output: sorted_reads/B_sorted.bam<br>    jobid: 0<br>    wildcards: sample=B<br><br>samtools sort -T sorted_reads/B -O bam mapped_reads/B.bam &gt; sorted_reads/B_sorted.bam<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        1       bwa_map<br>        1       samtools_sort<br>        2<br>This was a dry-run (flag -n). The order of <span class="hljs-built_in">jobs</span> does not reflect the order of execution.<br></code></pre></td></tr></table></figure>
<p>可以看到当指定输出为<code>B_sorted.bam</code>的时候,会先运行bwa得到<code>B.bam</code>然后再运行samtools得到<code>B_sorted.bam</code></p>
<h3 id="第四步：Indexing-read-alignments"><a href="#第四步：Indexing-read-alignments" class="headerlink" title="第四步：Indexing read alignments"></a>第四步：Indexing read alignments</h3><p>接下来我们需要使用<code>samtools</code>对排序的read alignments建立索引,将下面的规则补充到之前的规则下面:</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs python">rule samtools_index:<br>    <span class="hljs-built_in">input</span>:<br>        <span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;_sorted.bam&quot;</span><br>    output:<br>        <span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;.bam.bai&quot;</span><br>    shell:<br>        <span class="hljs-string">&quot;samtools index &#123;input&#125; &#123;output&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<p>snakemake会将不同的任务串成有向无环图(DAG),可以使用下面的命令来可视化流程：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake --dag sorted_reads/&#123;A,B&#125;.bam.bai | dot -Tsvg &gt; dag.svg<br></code></pre></td></tr></table></figure>
<center>

<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/dag.svg" srcset="/img/loading.gif" lazyload></p>
</center>

<p>DAG图的节点表示任务,边表示任务间的依赖关系,在节点中还会展示通配符的值(如<code>sample:B</code>);另外不需要运行的规则就用虚线边框表示：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment">##运行第一个规则</span><br>snakemake --cores 1 mapped_reads/A.bam<br><br><span class="hljs-comment">##再创建DAG</span><br>snakemake --dag sorted_reads/&#123;A,B&#125;.bam.bai | dot -Tsvg &gt; dag1.svg<br></code></pre></td></tr></table></figure>
<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/dag1.svg" srcset="/img/loading.gif" lazyload></p>
<h3 id="第五步：Calling-genomic-variants"><a href="#第五步：Calling-genomic-variants" class="headerlink" title="第五步：Calling genomic variants"></a>第五步：Calling genomic variants</h3><p>接下来需要检测reads中的变异信息,用的工具为：<code>samtools</code>和<code>bcftools</code></p>
<p>需要注意的是<code>snakemake</code>是通过目标文件(输出)来反推输入文件的,上面几个规则中的输出和输入共用一个通配符(sample);而这个步骤的输入是所有的bam及其索引,输出是一个文件(vcf)，所以需要在input指令下将所有的输入文件都写出来(因为此时snakemake无法通过输出推断输入)</p>
<p>snakemake提供了一个<code>expand</code>函数,可以方便的将文件名收集起来,就不需要一个一个写了：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs python">SAMPLES = [<span class="hljs-string">&quot;A&quot;</span>, <span class="hljs-string">&quot;B&quot;</span>]<br><br>expand(<span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;.bam&quot;</span>, sample=SAMPLES)<br><br></code></pre></td></tr></table></figure>
<p>将<code>SAMPLE</code>列表中的内容取代前面的通配符,也可以提供多个通配符,得到的结果是多个通配符的乘积：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs python">expand(<span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;.&#123;replicate&#125;.bam&quot;</span>, sample=SAMPLES, replicate=[<span class="hljs-number">0</span>, <span class="hljs-number">1</span>])<br><br><span class="hljs-comment">###结果是：</span><br>[<span class="hljs-string">&quot;sorted_reads/A.0.bam&quot;</span>, <span class="hljs-string">&quot;sorted_reads/A.1.bam&quot;</span>, <span class="hljs-string">&quot;sorted_reads/B.0.bam&quot;</span>, <span class="hljs-string">&quot;sorted_reads/B.1.bam&quot;</span>]<br></code></pre></td></tr></table></figure>

<p>因此需要在<code>Snakefile</code>的最前面定义<code>SAMPLES</code>,然后将下面的规则放到之前的规则下面：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs python">SAMPLES = [<span class="hljs-string">&quot;A&quot;</span>, <span class="hljs-string">&quot;B&quot;</span>]<br><br>...<br><br><br>rule bcftools_call:<br>    <span class="hljs-built_in">input</span>:<br>        fa=<span class="hljs-string">&quot;data/genome.fa&quot;</span>,<br>        bam=expand(<span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;_sorted.bam&quot;</span>, sample=SAMPLES),<br>        bai=expand(<span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;.bam.bai&quot;</span>, sample=SAMPLES)<br>    output:<br>        <span class="hljs-string">&quot;calls/all.vcf&quot;</span><br>    shell:<br>        <span class="hljs-string">&quot;samtools mpileup -g -f &#123;input.fa&#125; &#123;input.bam&#125; | &quot;</span><br>        <span class="hljs-string">&quot;bcftools call -mv - &gt; &#123;output&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<p>在shell命令中可以通过名称或者位置来指定输入或输出文件：</p>
<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/image-20210321171336544.png" srcset="/img/loading.gif" lazyload></p>
<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/image-20210321171348704.png" srcset="/img/loading.gif" lazyload></p>
<p>另外一个需要注意的是：如果命令太长,可以分多行写,但是在每行的命令的末尾要留一个空格:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment">##如果不留空格,拼起来会出错</span><br><span class="hljs-string">&quot;samtools mpileup&quot;</span><br><span class="hljs-string">&quot;-g -f &#123;input.fa&#125; &#123;input.bam&#125;&quot;</span><br><br><span class="hljs-string">&quot;samtools mpileup-g -f &#123;input.fa&#125; &#123;input.bam&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<p>现在再来看一下DAG图：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake --dag calls/all.vcf | dot -Tsvg &gt; dag2.svg<br></code></pre></td></tr></table></figure>

<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/dag2.svg" srcset="/img/loading.gif" lazyload></p>
<h3 id="第六步：Using-custom-scripts"><a href="#第六步：Using-custom-scripts" class="headerlink" title="第六步：Using custom scripts"></a>第六步：Using custom scripts</h3><p>在Snakemake中还可以引用外部的脚本(python/R)来完成一系列的工作;需要在<code>script</code>指令中指定脚本的路径:</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs python">rule plot_quals:<br>    <span class="hljs-built_in">input</span>:<br>        <span class="hljs-string">&quot;calls/all.vcf&quot;</span><br>    output:<br>        <span class="hljs-string">&quot;plots/quals.svg&quot;</span><br>    script:<br>        <span class="hljs-string">&quot;scripts/plot-quals.py&quot;</span><br></code></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">##plot-quals.py</span><br><span class="hljs-keyword">import</span> matplotlib<br>matplotlib.use(<span class="hljs-string">&quot;Agg&quot;</span>)<br><span class="hljs-keyword">import</span> matplotlib.pyplot <span class="hljs-keyword">as</span> plt<br><span class="hljs-keyword">from</span> pysam <span class="hljs-keyword">import</span> VariantFile<br><br>quals = [record.qual <span class="hljs-keyword">for</span> record <span class="hljs-keyword">in</span> VariantFile(snakemake.<span class="hljs-built_in">input</span>[<span class="hljs-number">0</span>])]<br>plt.hist(quals)<br><br>plt.savefig(snakemake.output[<span class="hljs-number">0</span>])<br></code></pre></td></tr></table></figure>
<p>上面的规则和python脚本最终产生一个变异检测质量值的直方图</p>
<p>规则的不同部分(input,output,wildcards等)在外部脚本中都是<code>snakemake</code>对象的属性,比如<code>snakemake.input</code>就是含有输入文件名的列表</p>
<p>同样在R代码中,<code>snakemake</code>是作为S4对象存在的,S4类的属性是R列表,因此我们可以通过类似<code>snakemake@input[[1]]</code>的形式来获取第一个输入文件</p>
<h3 id="第七步：Adding-a-target-rule"><a href="#第七步：Adding-a-target-rule" class="headerlink" title="第七步：Adding a target rule"></a>第七步：Adding a target rule</h3><p>前面都是通过命令行来指定目标文件(snakemake是通过目标文件逐步反推),如果没有指定目标文件则认为第一个规则的output是目标文件,因此我们可以在第一个规则中不加output,只包含input,并且这个input是整个流程的最终输出文件,这样我们就可以无需指定输出文件了(如果输出有很多,这样就不方便)：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">##将下面的规则放到第一个</span><br>rule <span class="hljs-built_in">all</span>:<br>    <span class="hljs-built_in">input</span>:<br>        <span class="hljs-string">&quot;plots/quals.svg&quot;</span><br></code></pre></td></tr></table></figure>

<p>可以看看整个流程的DAG图：</p>
<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/dag3.svg" srcset="/img/loading.gif" lazyload></p>
<p>接下来运行整个流程：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake --cores 1<br></code></pre></td></tr></table></figure>
<p>得到最终的结果为变异检测质量值的直方图：</p>
<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/quals.svg" srcset="/img/loading.gif" lazyload></p>
<p>注意：snakemake只在以下几种情况下才会执行任务(jobs)<br><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/image-20210322103940316.png" srcset="/img/loading.gif" lazyload></p>
<p>但是<code>Snakemake</code>也提供了强制运行的选项<code>--forcerun</code>,后面接输出文件或者规则(<code>--forceall</code>可以强制运行整个流程)：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br></pre></td><td class="code"><pre><code class="hljs bash">snakemake --cores 1 --forcerun samtools_sort<br>Building DAG of <span class="hljs-built_in">jobs</span>...<br>Using shell: /usr/bin/bash<br>Provided cores: 1 (use --cores to define parallelism)<br>Rules claiming more threads will be scaled down.<br>Job counts:<br>        count   <span class="hljs-built_in">jobs</span><br>        1       att<br>        1       bcftools_call<br>        1       plot_quals<br>        2       samtools_index<br>        2       samtools_sort<br>        7<br>Select <span class="hljs-built_in">jobs</span> to execute...<br><br>[Mon Mar 22 10:29:47 2021]<br>rule samtools_sort:<br>    input: mapped_reads/A.bam<br>    output: sorted_reads/A_sorted.bam<br>    jobid: 3<br>    wildcards: sample=A<br><br>[Mon Mar 22 10:29:48 2021]<br>Finished job 3.<br>1 of 7 steps (14%) <span class="hljs-keyword">done</span><br>Select <span class="hljs-built_in">jobs</span> to execute...<br><br>[Mon Mar 22 10:29:48 2021]<br>rule samtools_index:<br>    input: sorted_reads/A_sorted.bam<br>    output: sorted_reads/A.bam.bai<br>    jobid: 7<br>    wildcards: sample=A<br><br>[Mon Mar 22 10:29:48 2021]<br>Finished job 7.<br>2 of 7 steps (29%) <span class="hljs-keyword">done</span><br>Select <span class="hljs-built_in">jobs</span> to execute...<br><br>[Mon Mar 22 10:29:48 2021]<br>rule samtools_sort:<br>    input: mapped_reads/B.bam<br>    output: sorted_reads/B_sorted.bam<br>    jobid: 5<br>    wildcards: sample=B<br><br>[Mon Mar 22 10:29:48 2021]<br>Finished job 5.<br>3 of 7 steps (43%) <span class="hljs-keyword">done</span><br>Select <span class="hljs-built_in">jobs</span> to execute...<br><br>[Mon Mar 22 10:29:48 2021]<br>rule samtools_index:<br>    input: sorted_reads/B_sorted.bam<br>    output: sorted_reads/B.bam.bai<br>    jobid: 8<br>    wildcards: sample=B<br><br>[Mon Mar 22 10:29:48 2021]<br>Finished job 8.<br>4 of 7 steps (57%) <span class="hljs-keyword">done</span><br>Select <span class="hljs-built_in">jobs</span> to execute...<br><br>[Mon Mar 22 10:29:48 2021]<br>rule bcftools_call:<br>    input: data/genome.fa, sorted_reads/A_sorted.bam, sorted_reads/B_sorted.bam, sorted_reads/A.bam.bai, sorted_reads/B.bam.bai<br>    output: calls/all.vcf<br>    jobid: 2<br><br>[warning] samtools mpileup option `g` is functional, but deprecated. Please switch to using bcftools mpileup <span class="hljs-keyword">in</span> future.<br>[mpileup] 2 samples <span class="hljs-keyword">in</span> 2 input files<br>Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid<br>[Mon Mar 22 10:29:49 2021]<br>Finished job 2.<br>5 of 7 steps (71%) <span class="hljs-keyword">done</span><br>Select <span class="hljs-built_in">jobs</span> to execute...<br><br>[Mon Mar 22 10:29:49 2021]<br>rule plot_quals:<br>    input: calls/all.vcf<br>    output: plots/quals.svg<br>    jobid: 1<br><br>[Mon Mar 22 10:30:00 2021]<br>Finished job 1.<br>6 of 7 steps (86%) <span class="hljs-keyword">done</span><br>Select <span class="hljs-built_in">jobs</span> to execute...<br><br>[Mon Mar 22 10:30:00 2021]<br>localrule att:<br>    input: plots/quals.svg<br>    jobid: 0<br><br>[Mon Mar 22 10:30:00 2021]<br>Finished job 0.<br>7 of 7 steps (100%) <span class="hljs-keyword">done</span><br>Complete <span class="hljs-built_in">log</span>: /slst/home/wutao2/snakemake_tutorial/.snakemake/<span class="hljs-built_in">log</span>/2021-03-22T102940.208336.snakemake.log<br></code></pre></td></tr></table></figure>

<h2 id="进阶用法"><a href="#进阶用法" class="headerlink" title="进阶用法"></a>进阶用法</h2><h3 id="第一步：指定使用的线程数"><a href="#第一步：指定使用的线程数" class="headerlink" title="第一步：指定使用的线程数"></a>第一步：指定使用的线程数</h3><p>可以在规则中使用<code>threads</code>指令来指定需要的线程数(实际用到的可以小于等于指定的线程数),比如可以把bwa的线程指定为8个(如果不指定,默认是1)：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs python">rule bwa_map:<br>    <span class="hljs-built_in">input</span>:<br>        <span class="hljs-string">&quot;data/genome.fa&quot;</span>,<br>        <span class="hljs-string">&quot;data/samples/&#123;sample&#125;.fastq&quot;</span><br>    output:<br>        <span class="hljs-string">&quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br>    threads: <span class="hljs-number">8</span><br>    shell:<br>        <span class="hljs-string">&quot;bwa mem -t &#123;threads&#125; &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<p>在实际运行中用到的线程数由Snakemake来控制,保证同时运行的所有任务的总线程数不超过给定的总线程数,可以通过snakemake的参数<code>--cores</code>来指定给定的线程数(也就是说使用的线程数不超过<code>--cores</code>指定的数量),如果<code>--cores</code>后面没有数字则使用所有可用的核</p>
<h3 id="第二步：配置文件"><a href="#第二步：配置文件" class="headerlink" title="第二步：配置文件"></a>第二步：配置文件</h3><p>在前面的步骤中是通过一个python列表来指定需要考虑的样本(<code>SAMPLES = [&quot;A&quot;, &quot;B&quot;]</code>),但是如果想要流程能够更好的适应新的数据,我们可以使用配置文件,配置文件的格式可以是<code>JSON</code>或者<code>YAML</code>;在流程中使用<code>configfile</code>指令来指定配置文件</p>
<p><code>snakemake</code>会将配置文件读入,并将其内容存到名称为<code>config</code>的<strong>字典变量</strong>中</p>
<p>现在可以将之前的<code>SAMPLES</code>移除,加上配置文件：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">###配置文件config.yaml</span><br>samples:<br>    A: data/samples/A.fastq<br>    B: data/samples/B.fastq<br></code></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><code class="hljs python"><span class="hljs-comment">###将下面的指令放到Snakefile的开头</span><br>configfile: <span class="hljs-string">&quot;config.yaml&quot;</span><br><br><span class="hljs-comment">###将expand函数改写</span><br>rule bcftools_call:<br>    <span class="hljs-built_in">input</span>:<br>        fa=<span class="hljs-string">&quot;data/genome.fa&quot;</span>,<br>        bam=expand(<span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;_sorted.bam&quot;</span>, sample=config[<span class="hljs-string">&quot;samples&quot;</span>]),<span class="hljs-comment">##config是一个字典</span><br>        bai=expand(<span class="hljs-string">&quot;sorted_reads/&#123;sample&#125;.bam.bai&quot;</span>, sample=config[<span class="hljs-string">&quot;samples&quot;</span>])<br>    output:<br>        <span class="hljs-string">&quot;calls/all.vcf&quot;</span><br>    shell:<br>        <span class="hljs-string">&quot;samtools mpileup -g -f &#123;input.fa&#125; &#123;input.bam&#125; | &quot;</span><br>        <span class="hljs-string">&quot;bcftools call -mv - &gt; &#123;output&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<h3 id="第三步：输入函数"><a href="#第三步：输入函数" class="headerlink" title="第三步：输入函数"></a>第三步：输入函数</h3><p>上面通过<code>expand</code>和配置文件改写了<code>bcftools_call</code>规则,注意到配置文件中也有<code>fastq</code>文件的路径,我们可不可以也将<code>bwa_map</code>进行类似的改写呢？</p>
<p>首先需要了解Snakemake流程执行的步骤：</p>
<ul>
<li>初始化(initialization)：在流程内定义的文件被解析,所有的规则被实例化(<code>expand</code>函数就是在此时被执行)</li>
<li>DAG：通过填充通配符和根据输出文件匹配输入文件来构建任务的有向无环图</li>
<li>scheduling：根据可获得的资源来执行任务</li>
</ul>
<p>这两个规则的区别在于：<code>bcftools_call</code>不需要根据输出文件来推断输入文件(因为该步骤输入的是所有文件),而<code>bwa_map</code>规则需要根据输出来推断输入文件(比如,如果job的输出是B.bam,那么输入必须是B.fastq),因此无法在<code>bwa_map</code>中使用<code>expand</code>函数</p>
<p>但是我们可以使用<strong>input function</strong>来完成根据输出匹配输入的任务：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs python">rule bwa_map:<br>    <span class="hljs-built_in">input</span>:<br>        <span class="hljs-string">&quot;data/genome.fa&quot;</span>,<br>        <span class="hljs-keyword">lambda</span> wildcards: config[<span class="hljs-string">&quot;samples&quot;</span>][wildcards.sample]<br>    output:<br>        <span class="hljs-string">&quot;mapped_reads/&#123;sample&#125;.bam&quot;</span><br>    threads: <span class="hljs-number">8</span><br>    shell:<br>        <span class="hljs-string">&quot;bwa mem -t &#123;threads&#125; &#123;input&#125; | samtools view -Sb - &gt; &#123;output&#125;&quot;</span><br></code></pre></td></tr></table></figure>

<p>上面所展示的<code>input function</code>是lambda函数(为了方便所以使用lambda,一般的函数都可以的),参数是<code>wildcards</code>对象,通过<code>sample</code>属性获取output中匹配的通配符(比如output中是A.bam,wildcards.sample得到的就是”A”,<code>config[&quot;samples&quot;][wildcards.sample]</code>得到的就是配置文件中A的路径)</p>
<h3 id="第四步：规则参数"><a href="#第四步：规则参数" class="headerlink" title="第四步：规则参数"></a>第四步：规则参数</h3><p>有时候shell命令中的参数并不是静态的,比如要根据输入的样本名调整某些参数的值;snakemake提供了<code>params</code>指令</p>

            </div>
            <hr>
            <div>
              <div class="post-metas mb-3">
                
                  <div class="post-meta mr-3">
                    <i class="iconfont icon-category"></i>
                    
                      <a class="hover-with-bg" href="/categories/python/">python</a>
                    
                      <a class="hover-with-bg" href="/categories/python/skills/">skills</a>
                    
                  </div>
                
                
                  <div class="post-meta">
                    <i class="iconfont icon-tags"></i>
                    
                      <a class="hover-with-bg" href="/tags/bioinformatics/">bioinformatics</a>
                    
                  </div>
                
              </div>
              
                <p class="note note-warning">
                  
                    本博客所有文章除特别声明外，均采用 <a target="_blank" href="https://creativecommons.org/licenses/by-sa/4.0/deed.zh" rel="nofollow noopener noopener">CC BY-SA 4.0 协议</a> ，转载请注明出处！
                  
                </p>
              
              
                <div class="post-prevnext">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/2021/03/25/neo_pred/">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">使用NeoPredPipe预测新抗原</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/2021/03/14/perceptron/">
                        <span class="hidden-mobile">感知机模型</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>
    
      <div class="d-none d-lg-block col-lg-2 toc-container" id="toc-ctn">
        <div id="toc">
  <p class="toc-header"><i class="iconfont icon-list"></i>&nbsp;目录</p>
  <div class="toc-body" id="toc-body"></div>
</div>

      </div>
    
  </div>
</div>

<!-- Custom -->


    

    
      <a id="scroll-top-button" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v"
                 for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>
    

    
  </main>

  <footer class="text-center mt-5 py-3">
  <div class="footer-content">
     <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
  </div>
  

  

  
</footer>


  <!-- SCRIPTS -->
  
  <script  src="https://cdn.jsdelivr.net/npm/nprogress@0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/nprogress@0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" ></script>
<script  src="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>

<!-- Plugins -->


  
    <script  src="/js/img-lazyload.js" ></script>
  



  



  <script  src="https://cdn.jsdelivr.net/npm/tocbot@4.12.2/dist/tocbot.min.js" ></script>



  <script  src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.5.7/dist/jquery.fancybox.min.js" ></script>



  <script  src="https://cdn.jsdelivr.net/npm/anchor-js@4.3.0/anchor.min.js" ></script>



  <script defer src="https://cdn.jsdelivr.net/npm/clipboard@2.0.8/dist/clipboard.min.js" ></script>






  <script  src="https://cdn.jsdelivr.net/npm/typed.js@2.0.11/lib/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var title = document.getElementById('subtitle').title;
      
      typing(title)
      
    })(window, document);
  </script>



  <script  src="/js/local-search.js" ></script>
  <script>
    (function () {
      var path = "/local-search.xml";
      $('#local-search-input').on('click', function() {
        searchFunc(path, 'local-search-input', 'local-search-result');
      });
      $('#modalSearch').on('shown.bs.modal', function() {
        $('#local-search-input').focus();
      });
    })()
  </script>





  

  
    <!-- MathJax -->
    <script>
      MathJax = {
        tex: {
          inlineMath: [['$', '$'], ['\\(', '\\)']]
        },
        options: {
          renderActions: {
            findScript: [10, doc => {
              document.querySelectorAll('script[type^="math/tex"]').forEach(node => {
                const display = !!node.type.match(/; *mode=display/);
                const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display);
                const text = document.createTextNode('');
                node.parentNode.replaceChild(text, node);
                math.start = { node: text, delim: '', n: 0 };
                math.end = { node: text, delim: '', n: 0 };
                doc.math.push(math);
              });
            }, '', false],
            insertedScript: [200, () => {
              document.querySelectorAll('mjx-container').forEach(node => {
                let target = node.parentNode;
                if (target.nodeName.toLowerCase() === 'li') {
                  target.parentNode.classList.add('has-jax');
                }
              });
            }, '', false]
          }
        }
      };
    </script>

    <script async src="https://cdn.jsdelivr.net/npm/mathjax@3.1.2/es5/tex-svg.js" ></script>

  











<!-- 主题的启动项 保持在最底部 -->
<script  src="/js/boot.js" ></script>


</body>
</html>
