---
title: ä½¿ç”¨RåŒ…GenomicDataCommonsä¸‹è½½å’Œå¤„ç†TCGAæ•°æ®
author: wutao
date: 2022-01-07 10:00:00
categories:
  - ç”Ÿç‰©ä¿¡æ¯
index_img: img/GenomicDataCommons.png

---



GenomicDataCommons R åŒ…å­¦ä¹ ï¼Œå¹¶ä½¿ç”¨è¯¥åŒ…è®¡ç®— TCGA æ ·æœ¬çš„æµ‹åºæ·±åº¦
<!-- more -->

GDC (Genomic Data Commons) æ˜¯ç¾å›½å›½å®¶ç™Œç—‡ç ”ç©¶æ‰€å»ºç«‹çš„åœ¨ç™Œç—‡ç²¾å‡†åŒ»ç–—æ•°æ®æ–¹é¢çš„æ•°æ®å…±äº«å¹³å°ï¼Œç›®å‰å·²ç»åŒ…å«å‡ ä¸ªå¤§å‹çš„ç™Œç—‡åŸºå› ç»„æ•°æ®é›†ï¼Œæ¯”å¦‚ TCGA å’Œ TARGETã€‚GDC çš„æ•°æ®æ¨¡å‹éå¸¸å¤æ‚ï¼Œå¯ä»¥ç”¨ä¸‹å›¾æ¥ç®€å•å±•ç¤ºï¼š

![](https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/%E4%B8%8B%E8%BD%BD.png)

èŠ‚ç‚¹è¡¨ç¤ºå®ä½“ï¼Œæ¯”å¦‚é¡¹ç›®ï¼Œæ ·æœ¬ï¼Œè¯Šæ–­ï¼Œæ–‡ä»¶ç­‰ï¼Œå®ä½“ä¹‹é—´çš„å…³ç³»ç”¨è¾¹æ¥è¡¨ç¤ºï¼ŒèŠ‚ç‚¹å’Œè¾¹éƒ½æœ‰ç›¸åº”çš„å±æ€§ã€‚

# å¿«é€Ÿå¼€å§‹

è¿™ä¸€éƒ¨åˆ†å±•ç¤ºä¸€äº›åŸºç¡€åŠŸèƒ½

## å®‰è£…

å®‰è£…å’Œä¸€èˆ¬çš„ `Bioconductor` åŒ…ä¸€æ ·ï¼š

``` r
if (!require("BiocManager"))
    install.packages("BiocManager")
BiocManager::install('GenomicDataCommons')
```

``` r
library(GenomicDataCommons)
```

## æ£€æŸ¥è¿æ¥å’ŒçŠ¶æ€

`GenomicDataCommons` åŒ…éœ€è¦ç½‘ç»œè¿æ¥ï¼Œå¹¶ä¸”åœ¨ä½¿ç”¨æ—¶ NCI çš„ GDC API å¤„äºå¯æ“ä½œå’Œéç»´æŠ¤çŠ¶æ€ï¼Œä½¿ç”¨ `status` æ¥æ£€æŸ¥è¿æ¥å’ŒçŠ¶æ€ï¼š

``` r
GenomicDataCommons::status()
>> $commit
>> [1] "b49b90e1318040f447906940f3dff145809d9ea0"
>> 
>> $data_release
>> [1] "Data Release 31.0 - October 29, 2021"
>> 
>> $status
>> [1] "OK"
>> 
>> $tag
>> [1] "3.0.0"
>> 
>> $version
>> [1] 1
```

å¦‚æœæˆ‘ä»¬éœ€è¦åœ¨è„šæœ¬æˆ–è€…å¼€å‘çš„åŒ…ä¸­åˆ¤æ–­è¿æ¥æ˜¯å¦æ­£å¸¸ï¼Œå¯ä»¥ä½¿ç”¨ `stopifnot` å‡½æ•°ï¼š

``` r
stopifnot(GenomicDataCommons::status()$status=="OK")
```

## å¯»æ‰¾æ•°æ®

åœ¨ä¸‹è½½æ•°æ®ä¹‹å‰ï¼Œæˆ‘ä»¬éœ€è¦å…ˆåˆ¶ä½œåŸå§‹æ•°æ®çš„ `manifest` æ–‡ä»¶ï¼Œè¿™ä¸ªæ–‡ä»¶ä¸­æœ‰æ•°æ®çš„ UUID ï¼Œå¯ä»¥è¢« GDC çš„ API
ç”¨æ¥å®šä½ä¸‹è½½çš„æ–‡ä»¶ã€‚æ¯”å¦‚ä¸‹é¢çš„ä»£ç è·å–äº†åµå·¢ç™Œ RNA-seqæ•°æ®çš„åŸå§‹ countsï¼š

``` r
ge_manifest = files() %>%
    filter( cases.project.project_id == 'TCGA-OV') %>% 
    filter( type == 'gene_expression' ) %>%
    filter( analysis.workflow_type == 'HTSeq - Counts')  %>%
    manifest()
>> Rows: 379 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
head(ge_manifest)
>> # A tibble: 6 Ã— 5
>>   id                                   filename                                             md5                      size state 
>>   <chr>                                <chr>                                                <chr>                   <dbl> <chr> 
>> 1 451c9c6d-f062-4ca2-8459-efed929bd305 9c6f2d76-7701-434e-a876-c04ab14cccea.htseq.counts.gz d296b0cb99d14b4b91383â€¦ 263970 releaâ€¦
>> 2 5bc23348-57f9-4929-a90b-a557696ca955 af6e5654-e755-4c15-b3e5-807da2642e25.htseq.counts.gz 615e3868d8943e94859a2â€¦ 253867 releaâ€¦
>> 3 7c8ea118-fbce-4dc9-803d-1b6e30d06704 d73c0f69-ab9b-4408-bf89-aa34bf829351.htseq.counts.gz ef4a8c5d45c6de49b590bâ€¦ 263239 releaâ€¦
>> 4 4b83b0f5-4fc7-4a3a-a090-3ce497fb1af4 c9689d9f-6138-42a8-a58e-1b44dc4b193f.htseq.counts.gz f1ffd9f86b0f6c97cbb66â€¦ 257423 releaâ€¦
>> 5 749e9e26-eb75-4681-b039-6966e911ae7a 43622957-7bed-4d24-a31e-5fadf41216e1.htseq.counts.gz 50e84291a89a72fe3d43eâ€¦ 255649 releaâ€¦
>> 6 a2d8e28d-0371-4421-8ca4-4629a8cc7b72 9a420e2d-15c8-41f3-859f-25872765f75e.htseq.counts.gz 3e2e347efa85410ecbd8câ€¦ 253306 releaâ€¦
nrow(ge_manifest)
>> [1] 379
```

## ä¸‹è½½æ•°æ®

é€šè¿‡ä¸Šé¢æ­¥éª¤çš„ç­›é€‰ï¼Œè·å¾—äº† 379 ä¸ªæ ·æœ¬çš„åŸºå› è¡¨è¾¾æ•°æ®çš„ UUIDï¼Œæˆ‘ä»¬å¯ä»¥é€šè¿‡è¿™äº› UUID æ¥ä¸‹è½½ç›¸åº”çš„åŸºå› è¡¨è¾¾æ•°æ®ï¼š

``` r
##æ¼”ç¤ºåªä¸‹è½½äº†å‰20ä¸ªæ–‡ä»¶ï¼Œå·²ç»æŒ‡å®šäº† cache dir
fnames = lapply(ge_manifest$id[1:20],gdcdata)
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
>> Rows: 1 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
```

ä¸‹è½½éœ€è¦ç”¨åˆ°çš„å‡½æ•°æ˜¯ `gdcdata`ï¼Œè¯¥å‡½æ•°æ¥å—æ–‡ä»¶çš„ UUID ä½œä¸ºè¾“å…¥ï¼Œè¿”å›ä¸‹è½½çš„æ–‡ä»¶å¹¶ä»¥ `Filename` é‡å‘½åæ–‡ä»¶ï¼Œéœ€è¦æ³¨æ„çš„æ˜¯æˆ‘ä»¬æœ€å¥½æŒ‡å®š`gdc_cache` ä½œä¸º ä¸‹è½½æ–‡ä»¶çš„å­˜æ”¾ç›®å½•ï¼ˆcacheï¼‰ï¼Œé¿å…å¤šæ¬¡é‡å¤ä¸‹è½½ï¼ˆå½“ç„¶é»˜è®¤çš„ cache ç›®å½•ä¹Ÿå¯ä»¥ï¼‰ã€‚

å¯ä»¥æŸ¥çœ‹ä¸€ä¸‹ä¸‹è½½æ–‡ä»¶çš„ç»“æ„ï¼š

``` bash
tree ~/.cache/GenomicDataCommons
>> [01;34m/home/data/t040201/.cache/GenomicDataCommons[00m
>> â”œâ”€â”€ [01;34m02ab2f95-023b-492e-8dc9-20b4f52f209f[00m
>> â”‚Â Â  â””â”€â”€ [01;31mb1d5c444-da0d-4360-bd45-31c94217adfc.FPKM-UQ.txt.gz[00m
>> â”œâ”€â”€ [01;34m0f2ef2fa-cf6e-4d03-a648-ccd9fdf8b8c5[00m
>> â”‚Â Â  â””â”€â”€ [01;31mb64ef80a-d41c-4f92-a3ed-e43d55abb2c2.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m1480554d-8579-4146-93ba-a7b23b8c9a5b[00m
>> â”‚Â Â  â””â”€â”€ [01;31ma2082ad4-279e-422f-b5a7-cb7fbeb7a6df.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m1c099182-703f-4aaf-b1d6-0f9be5094b9a[00m
>> â”‚Â Â  â””â”€â”€ [01;31mfea3c4d0-7b9f-4279-989e-535aaefbdfeb.FPKM.txt.gz[00m
>> â”œâ”€â”€ [01;34m3b82236e-b055-4d10-acb4-b5a5992a1261[00m
>> â”‚Â Â  â””â”€â”€ [01;31m6aa7225b-3d10-4b19-a472-0adeb21c26cf.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m451c9c6d-f062-4ca2-8459-efed929bd305[00m
>> â”‚Â Â  â””â”€â”€ [01;31m9c6f2d76-7701-434e-a876-c04ab14cccea.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m4958b9d4-59fa-49e4-a849-1b777452f6d2[00m
>> â”‚Â Â  â””â”€â”€ [01;31m3e00e8aa-31bc-454e-b558-7bcbad5f47ad.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m4b83b0f5-4fc7-4a3a-a090-3ce497fb1af4[00m
>> â”‚Â Â  â””â”€â”€ [01;31mc9689d9f-6138-42a8-a58e-1b44dc4b193f.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m581b0bcf-a7e5-44ec-818a-9193a0565095[00m
>> â”‚Â Â  â””â”€â”€ [01;31m347a8eeb-204f-41df-98d3-61394d2b7cd7.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m5bc23348-57f9-4929-a90b-a557696ca955[00m
>> â”‚Â Â  â””â”€â”€ [01;31maf6e5654-e755-4c15-b3e5-807da2642e25.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m5e57bd1d-f38d-447d-9d7c-1fa1c5a2ecaa[00m
>> â”‚Â Â  â””â”€â”€ [01;31mb269c35d-7f91-4c66-8bef-59906ec87745.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m717ca3c0-0b0e-4cf3-af34-15d0fbfd7f68[00m
>> â”‚Â Â  â””â”€â”€ [01;31mc255d022-a659-42bd-9099-6853b41b64c7.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m749e9e26-eb75-4681-b039-6966e911ae7a[00m
>> â”‚Â Â  â””â”€â”€ [01;31m43622957-7bed-4d24-a31e-5fadf41216e1.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m750f8926-4361-4a92-8c72-59c82baad867[00m
>> â”‚Â Â  â””â”€â”€ [01;31m195a13be-31a4-47ce-bc3d-6aba8451e304.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m769c5f60-fdf1-49b1-bba3-7f4a3ef1f9a8[00m
>> â”‚Â Â  â””â”€â”€ [01;31m9e198dfe-9fc6-48e2-ba06-90c49ddf48aa.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34m7c8ea118-fbce-4dc9-803d-1b6e30d06704[00m
>> â”‚Â Â  â””â”€â”€ [01;31md73c0f69-ab9b-4408-bf89-aa34bf829351.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34ma0765b46-5fbb-49d0-b8aa-682486927d0f[00m
>> â”‚Â Â  â””â”€â”€ [01;31m5a928267-356d-47e9-b8b2-f477eaa261fa.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34ma2d8e28d-0371-4421-8ca4-4629a8cc7b72[00m
>> â”‚Â Â  â””â”€â”€ [01;31m9a420e2d-15c8-41f3-859f-25872765f75e.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34mb44b0d52-562a-476a-a607-a7148d128359[00m
>> â”‚Â Â  â””â”€â”€ [01;31m062b6d0c-06b0-4d27-8702-32501278cd9c.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34mc44d4c5c-b855-4e82-bd90-0f4b6c0e0015[00m
>> â”‚Â Â  â””â”€â”€ [01;31m75303476-cdec-4ae4-aaf5-01abdc3213ab.htseq.counts.gz[00m
>> â”œâ”€â”€ [01;34mc9cd12a4-b4b3-416e-9046-9c06dd884547[00m
>> â”‚Â Â  â””â”€â”€ [01;31mac6e38b7-143c-491e-9892-4c28f51ddce5.htseq.counts.gz[00m
>> â””â”€â”€ [01;34mcd32019f-e6e5-4cf9-a4b4-d38dfdfc0f0a[00m
>>     â””â”€â”€ [01;31mfedd52be-18a8-423f-ba8a-4f9416f11ff5.htseq.counts.gz[00m
>> 
>> 22 directories, 22 files
```

<p class="note note-primary">
å¦‚æœä¸‹è½½çš„æ˜¯ `controlled-access` æ•°æ®ï¼Œéœ€è¦æä¾› `token`
</p>


## å…ƒæ•°æ®è·å–

### ä¸´åºŠæ•°æ®

è·å–ä¸´åºŠæ•°æ®æ˜¯ä¸€ä¸ªå¸¸è§çš„ä»»åŠ¡ï¼Œ`gdc_clinical` å‡½æ•°æ¥å— `case_ids`ï¼Œè¿”å›ä¸€ä¸ªæœ‰ä¸ªå››ä¸ªæ•°æ®æ¡†çš„åˆ—è¡¨ï¼š

-   äººå£å­¦ç»Ÿè®¡ï¼ˆdemographicï¼‰ï¼ŒåŒ…æ‹¬æ€§åˆ«ï¼Œç§æ—ï¼Œå¹´é¾„ç­‰
-   è¯Šæ–­ï¼ˆdiagnosesï¼‰ï¼ŒåŒ…æ‹¬ä¸´åºŠåˆ†æœŸï¼Œç”Ÿå­˜æ—¶é—´ï¼Œå›è®¿ä¿¡æ¯ç­‰
-   æš´éœ²ï¼ˆexposuresï¼‰ï¼ŒåŒ…æ‹¬å¸çƒŸï¼Œé¥®é…’è®°å½•ç­‰
-   mainï¼ŒåŒ…æ‹¬ç–¾ç—…ç±»å‹ï¼Œè¯Šæ–­æ—¶é—´ç­‰

``` r
case_ids = cases() %>% results(size=10) %>% ids()
clindat = gdc_clinical(case_ids)
names(clindat)
>> [1] "demographic" "diagnoses"   "exposures"   "main"

head(clindat[["demographic"]])
>> # A tibble: 6 Ã— 15
>>   vital_status gender race    ethnicity  age_at_index submitter_id  days_to_birth created_datetime year_of_birth demographic_id 
>>   <chr>        <chr>  <chr>   <chr>             <int> <chr>                 <int> <lgl>                    <int> <chr>          
>> 1 Alive        female not reâ€¦ not reporâ€¦           65 TCGA-A8-A07Gâ€¦        -23926 NA                        1944 38caac77-d856-â€¦
>> 2 Dead         female black â€¦ not hispaâ€¦           49 TCGA-A2-A3XYâ€¦        -18059 NA                        1961 ac6eba06-6e54-â€¦
>> 3 Alive        female white   not hispaâ€¦           56 TCGA-E2-A152â€¦        -20705 NA                        1953 680b1fdd-143c-â€¦
>> 4 Alive        female white   not hispaâ€¦           40 TCGA-E2-A15Eâ€¦        -14894 NA                        1969 1b53b23d-b98c-â€¦
>> 5 Alive        female white   not reporâ€¦           73 TCGA-AR-A0U0â€¦        -26993 NA                        1931 2f82f157-cce2-â€¦
>> 6 Alive        female white   not reporâ€¦           52 TCGA-BH-A0E1â€¦        -19192 NA                        1957 7c593975-bc80-â€¦
>> # â€¦ with 5 more variables: updated_datetime <chr>, state <chr>, year_of_death <lgl>, days_to_death <int>, case_id <chr>
head(clindat[["diagnoses"]])
>> # A tibble: 6 Ã— 29
>>   case_id synchronous_malâ€¦ ajcc_pathologicâ€¦ days_to_diagnosâ€¦ created_datetime last_known_diseâ€¦ tissue_or_organâ€¦ days_to_last_foâ€¦
>>   <chr>   <chr>            <chr>                       <int> <lgl>            <chr>            <chr>                       <int>
>> 1 8cf8b6â€¦ No               Stage IIA                       0 NA               not reported     Breast, NOS                   577
>> 2 deba32â€¦ No               Stage IIB                       0 NA               not reported     Breast, NOS                  1064
>> 3 a80154â€¦ No               Stage I                         0 NA               not reported     Breast, NOS                  2128
>> 4 0a2a35â€¦ No               Stage IIA                       0 NA               not reported     Breast, NOS                   630
>> 5 e3c336â€¦ No               Stage IIB                       0 NA               not reported     Breast, NOS                  1988
>> 6 606fbcâ€¦ No               Stage IIB                       0 NA               not reported     Breast, NOS                   477
>> # â€¦ with 21 more variables: primary_diagnosis <chr>, age_at_diagnosis <int>, updated_datetime <chr>, year_of_diagnosis <int>,
>> #   prior_malignancy <chr>, state <chr>, prior_treatment <chr>, days_to_last_known_disease_status <lgl>,
>> #   ajcc_staging_system_edition <chr>, ajcc_pathologic_t <chr>, days_to_recurrence <lgl>, morphology <chr>,
>> #   ajcc_pathologic_n <chr>, ajcc_pathologic_m <chr>, submitter_id <chr>, classification_of_tumor <chr>, diagnosis_id <chr>,
>> #   icd_10_code <chr>, site_of_resection_or_biopsy <chr>, tumor_grade <chr>, progression_or_recurrence <chr>
head(clindat[["exposures"]])
>> # A tibble: 6 Ã— 10
>>   case_id    cigarettes_per_dâ€¦ updated_datetime   alcohol_history exposure_id   submitter_id years_smoked state created_datetime
>>   <chr>      <lgl>             <chr>              <chr>           <chr>         <chr>        <lgl>        <chr> <lgl>           
>> 1 8cf8b620-â€¦ NA                2019-07-31T21:48:â€¦ Not Reported    a8be57c5-620â€¦ TCGA-A8-A07â€¦ NA           releâ€¦ NA              
>> 2 deba32e4-â€¦ NA                2019-07-31T21:29:â€¦ Not Reported    9aa5cc63-ceaâ€¦ TCGA-A2-A3Xâ€¦ NA           releâ€¦ NA              
>> 3 a8015490-â€¦ NA                2019-07-31T21:52:â€¦ Not Reported    134cee43-117â€¦ TCGA-E2-A15â€¦ NA           releâ€¦ NA              
>> 4 0a2a3529-â€¦ NA                2019-07-31T21:31:â€¦ Not Reported    fdf22b2e-0baâ€¦ TCGA-E2-A15â€¦ NA           releâ€¦ NA              
>> 5 e3c336f5-â€¦ NA                2019-07-31T15:38:â€¦ Not Reported    3d1c43ad-b1dâ€¦ TCGA-AR-A0Uâ€¦ NA           releâ€¦ NA              
>> 6 606fbc6a-â€¦ NA                2019-07-31T21:42:â€¦ Not Reported    1d844394-425â€¦ TCGA-BH-A0Eâ€¦ NA           releâ€¦ NA              
>> # â€¦ with 1 more variable: alcohol_intensity <lgl>
head(clindat[["main"]])
>> # A tibble: 6 Ã— 8
>>   id                                   disease_type  submitter_id created_datetime primary_site updated_datetime  case_id  state
>>   <chr>                                <chr>         <chr>        <lgl>            <chr>        <chr>             <chr>    <chr>
>> 1 8cf8b620-7ab6-4b6e-84bc-ff5a83f381fa Ductal and Lâ€¦ TCGA-A8-A07G NA               Breast       2019-08-06T14:14â€¦ 8cf8b62â€¦ releâ€¦
>> 2 deba32e4-0e68-4711-941b-3b63bd965afb Ductal and Lâ€¦ TCGA-A2-A3XY NA               Breast       2019-08-06T14:14â€¦ deba32eâ€¦ releâ€¦
>> 3 a8015490-9740-45c9-8bd2-eb6d1beefc2e Ductal and Lâ€¦ TCGA-E2-A152 NA               Breast       2019-08-06T14:16â€¦ a801549â€¦ releâ€¦
>> 4 0a2a3529-f645-4967-9a58-89ee20b8bb62 Ductal and Lâ€¦ TCGA-E2-A15E NA               Breast       2019-08-06T14:16â€¦ 0a2a352â€¦ releâ€¦
>> 5 e3c336f5-c32f-4c5d-81fb-e2408ae145b2 Ductal and Lâ€¦ TCGA-AR-A0U0 NA               Breast       2019-08-06T14:15â€¦ e3c336fâ€¦ releâ€¦
>> 6 606fbc6a-b41b-441d-9401-51e54912bf5e Ductal and Lâ€¦ TCGA-BH-A0E1 NA               Breast       2019-08-06T14:15â€¦ 606fbc6â€¦ releâ€¦
```

### å¹¿ä¹‰çš„å…ƒæ•°æ®è·å–

æˆ‘ä»¬å¯ä»¥é€šè¿‡ `GenomicDataCommons` ç»“åˆå„ç§å‚æ•°ï¼ˆæ¯”å¦‚ `filter`, `select`, `expand`ç­‰ï¼‰æ¥çµæ´»çš„è·å–æƒ³è¦çš„å…ƒæ•°æ®ï¼š

``` r
expands = c("diagnoses","annotations",
             "demographic","exposures")
clinResults = cases() %>%
    GenomicDataCommons::select(NULL) %>%
    GenomicDataCommons::expand(expands) %>%
    results(size=50)
str(clinResults[[1]],list.len=6)
>>  chr [1:50] "8cf8b620-7ab6-4b6e-84bc-ff5a83f381fa" "deba32e4-0e68-4711-941b-3b63bd965afb" ...
```

# åŸºæœ¬è®¾è®¡

ä»ä¸Šé¢çš„ä¾‹å­ä¸­å¯ä»¥çœ‹å‡ºè¿™ä¸ªåŒ…çš„è®¾è®¡è¡Œä¸ºå’Œ `dplyr` æ˜¯éå¸¸ç±»ä¼¼çš„ï¼Œä¸€äº›åŠ¨è¯ä¹Ÿæ˜¯ç›´æ¥å’Œ `dplyr` çš„å‡½æ•°åŒå(filter, select ç­‰)ã€‚ç®€å•æ¥è¯´ï¼Œè¯¥åŒ…æ£€ç´¢ï¼Œè·å–å…ƒæ•°æ®å’Œæ–‡ä»¶åˆ†ä¸ºä¸‰æ­¥ï¼š

-   è¯·æ±‚æ„é€ å‡½æ•°ï¼ˆquery constructorsï¼Œå¦‚ä¸Šé¢çš„ `cases()`, `files()`ï¼‰
-   ä¸€ç³»åˆ—çš„åŠ¨è¯ç”¨æ¥è¿‡æ»¤æ ·æœ¬ï¼Œé€‰æ‹©å­—æ®µï¼Œèšåˆå¹¶ç”Ÿæˆæœ€ç»ˆçš„è¯·æ±‚å¯¹è±¡ï¼ˆquery objectï¼Œå¦‚ `filter`, `select`ç­‰ï¼‰
-   ä½¿ç”¨ä¸€ç³»åˆ—åŠ¨è¯åŸºäºä¸Šé¢å¾—åˆ°çš„è¯·æ±‚å¯¹è±¡è·å–æ–‡ä»¶ï¼Œè¿”å›ç»“æœï¼ˆå¦‚ `gdcdata`ï¼‰

å®Œæˆä¸Šè¿°è¿‡ç¨‹çš„åŸºæœ¬å‡½æ•°å¦‚ä¸‹ï¼š

-   åˆ›å»ºè¯·æ±‚ï¼š
    -   `projects()`
    -   `cases()`
    -   `files()`
    -   `annotations()`
-   æ“ä½œè¯·æ±‚ï¼š
    -   `filter()`
    -   `facet()`
    -   `select()`
    -   `expand()`
-   GDC API å­—æ®µçš„å†…çœï¼ˆintrospection, å³æ”¯æŒå“ªäº›æŸ¥è¯¢ï¼Œæœ‰å“ªäº›å­—æ®µåŠå­—æ®µç±»å‹ç­‰ä¿¡æ¯ï¼‰ï¼š
    -   `mapping()`
    -   `available_fields()`
    -   `default_fields()`
    -   `grep_fields()`
    -   `field_picker()`
    -   `available_values()`
    -   `available_expand()`
-   æ‰§è¡Œ API è°ƒç”¨ï¼Œè·å–è¯·æ±‚çš„ç»“æœï¼š
    -   `results()`
    -   `count()`
    -   `response()`
-   åŸå§‹æ•°æ®ä¸‹è½½ï¼š
    -   `gdcdata()`
    -   `transfer()`
    -   `gdc_client()`
-   æ±‡æ€»ï¼Œèšåˆå­—æ®µå€¼ï¼ˆä¹Ÿå« facetingï¼‰ï¼š
    -   `aggregations()`
-   Control æ•°æ®çš„æˆæƒï¼ˆtokenï¼‰ï¼š
    -   `gdc_token()`
-   BAM æ–‡ä»¶åˆ‡ç‰‡ï¼š
    -   `slicing()`

# ç”¨æ³•

ä¸Šé¢é‚£äº›åŸºæœ¬å‡½æ•°æ„æˆäº†ä¸¤å¤§ç±»æ“ä½œï¼š

-   æ£€ç´¢å…ƒæ•°æ®å’Œå¯»æ‰¾æ•°æ®æ–‡ä»¶
-   ä¼ è¾“åŸå§‹æ•°æ®æˆ–è€…å¤„ç†åçš„æ•°æ®

## æ£€ç´¢å…ƒæ•°æ®

### åˆ›å»ºæ£€ç´¢

æœ‰å››ç§æ–¹ä¾¿çš„å‡½æ•°å¯ä»¥åˆ›å»º `GDCQuery` å¯¹è±¡æ¥è·å–ä¸åŒç±»å‹çš„æ•°æ®ï¼š

-   `project()`
-   `cases()`
-   `files()`
-   `annotations()`

è¿™äº›å¯¹è±¡éƒ½å«æœ‰ä¸‹åˆ—çš„ä¸€äº›å…ƒç´ ï¼š

-   å­—æ®µï¼ˆfieldsï¼‰ï¼šéœ€è¦ä¸‹è½½çš„å­—æ®µï¼Œå¦‚æœæ²¡æœ‰æŒ‡å®šå­—æ®µï¼Œå°±ä¼šä½¿ç”¨é»˜è®¤å­—æ®µæ¥å–å›æ•°æ®ï¼ˆé»˜è®¤å­—æ®µå¯ä»¥é€šè¿‡`default_fields()` æ¥æŸ¥çœ‹ï¼‰
-   è¿‡æ»¤å™¨ï¼ˆfiltersï¼‰ï¼šå«æœ‰è°ƒç”¨ `filter()` æ–¹æ³•åè·å¾—çš„ç»“æœï¼Œå¹¶ç”¨æ¥ç­›é€‰å–å›çš„æ•°æ®
-   facetsï¼šå½“è°ƒç”¨ `aggregations()` æ—¶å¯¹æ•°æ®æ±‡æ€»æ‰€éœ€çš„å­—æ®µå
-   å­˜æ¡£ï¼ˆarchiveï¼‰ï¼šå¯ä»¥æ˜¯ â€œdefaultâ€ æˆ–è€… `legacy` (legacy æ˜¯æ¯”è¾ƒè€çš„æ•°æ®)
-   tokenï¼šä¸‹è½½ control æ•°æ®çš„å‡­è¯ï¼Œå¯¹äºè·å–å…ƒæ•°æ®å¯ä»¥ä¸éœ€è¦ï¼Œåªéœ€åœ¨ä¸‹è½½çœŸæ­£æ•°æ®æ—¶æä¾›å³å¯

``` r
pquery = projects()
str(pquery)
>> List of 5
>>  $ fields : chr [1:10] "dbgap_accession_number" "disease_type" "intended_release_date" "name" ...
>>  $ filters: NULL
>>  $ facets : NULL
>>  $ legacy : logi FALSE
>>  $ expand : NULL
>>  - attr(*, "class")= chr [1:3] "gdc_projects" "GDCQuery" "list"

default_fields("projects")
>>  [1] "dbgap_accession_number" "disease_type"           "intended_release_date"  "name"                  
>>  [5] "primary_site"           "project_autocomplete"   "project_id"             "releasable"            
>>  [9] "released"               "state"
```

å¯ä»¥çœ‹åˆ°åˆå§‹çŠ¶æ€ä¸‹å¤§éƒ¨åˆ†å…ƒç´ æ˜¯ç©ºçš„ï¼ˆNULLï¼‰

### å–å›ç»“æœ

å½“æœ‰äº†ä¸€ä¸ªè¯·æ±‚å¯¹è±¡åï¼Œæˆ‘ä»¬å°±å¯ä»¥ä» GDC å–å›ç»“æœäº†ã€‚å¯ä»¥ä½¿ç”¨ `count()` æ¥å¾—åˆ°æœ€åŸºæœ¬çš„ç»“æœç±»å‹ï¼Œè¿™ä¸ªå‡½æ•°è¿”å›æ»¡è¶³ `filter` æ ‡å‡†çš„è®°å½•æ•°é‡ï¼Œç”±äºæˆ‘ä»¬ç›®å‰æ²¡æœ‰å®šä¹‰ä»»ä½•çš„è¿‡æ»¤æ¡ä»¶ï¼Œæ‰€ä»¥è¿™é‡Œçš„ `count()` è¿”å›çš„æ˜¯æ‰€æœ‰çš„ `project` è®°å½•ï¼ˆåœ¨ default å­˜æ¡£ä¸­ï¼‰ï¼š

``` r
pcount = pquery %>% count()
pcount
>> [1] 70
```

`results()` å¯ä»¥ç›´æ¥å–å›ç»“æœï¼š

``` r
presults = pquery %>% results()
```

è¿”å›çš„ç»“æœä» GDC çš„ Json æ ¼å¼è¢«è½¬æ¢æˆ R é‡Œé¢çš„ Listï¼Œå¯ä»¥ä½¿ç”¨ `str()` æ¥ç®€å•çš„æŸ¥çœ‹æ•°æ®ç»“æ„ï¼š

``` r
str(presults)
>> List of 9
>>  $ id                    : chr [1:10] "TCGA-BRCA" "GENIE-MSK" "GENIE-VICC" "GENIE-UHN" ...
>>  $ primary_site          :List of 10
>>   ..$ TCGA-BRCA            : chr "Breast"
>>   ..$ GENIE-MSK            : chr [1:49] "Connective, subcutaneous and other soft tissues" "Kidney" "Prostate gland" "Other and unspecified major salivary glands" ...
>>   ..$ GENIE-VICC           : chr [1:46] "Connective, subcutaneous and other soft tissues" "Kidney" "Prostate gland" "Other and unspecified major salivary glands" ...
>>   ..$ GENIE-UHN            : chr [1:42] "Connective, subcutaneous and other soft tissues" "Kidney" "Prostate gland" "Other and unspecified major salivary glands" ...
>>   ..$ CPTAC-2              : chr [1:6] "Ovary" "Rectum" "Other and unspecified female genital organs" "Breast" ...
>>   ..$ CMI-ASC              : chr [1:9] "Bladder" "Other and ill-defined sites" "Other and ill-defined digestive organs" "Heart, mediastinum, and pleura" ...
>>   ..$ BEATAML1.0-COHORT    : chr "Hematopoietic and reticuloendothelial systems"
>>   ..$ CGCI-BLGSP           : chr "Hematopoietic and reticuloendothelial systems"
>>   ..$ BEATAML1.0-CRENOLANIB: chr "Hematopoietic and reticuloendothelial systems"
>>   ..$ CMI-MPC              : chr [1:2] "Prostate gland" "Lymph nodes"
>>  $ dbgap_accession_number: chr [1:10] NA NA NA NA ...
>>  $ project_id            : chr [1:10] "TCGA-BRCA" "GENIE-MSK" "GENIE-VICC" "GENIE-UHN" ...
>>  $ disease_type          :List of 10
>>   ..$ TCGA-BRCA            : chr [1:9] "Complex Epithelial Neoplasms" "Fibroepithelial Neoplasms" "Adnexal and Skin Appendage Neoplasms" "Adenomas and Adenocarcinomas" ...
>>   ..$ GENIE-MSK            : chr [1:49] "Miscellaneous Bone Tumors" "Myeloid Leukemias" "Gliomas" "Lipomatous Neoplasms" ...
>>   ..$ GENIE-VICC           : chr [1:43] "Leukemias, NOS" "Chronic Myeloproliferative Disorders" "Myeloid Leukemias" "Gliomas" ...
>>   ..$ GENIE-UHN            : chr [1:39] "Leukemias, NOS" "Chronic Myeloproliferative Disorders" "Miscellaneous Bone Tumors" "Myeloid Leukemias" ...
>>   ..$ CPTAC-2              : chr [1:5] "Not Reported" "Cystic, Mucinous and Serous Neoplasms" "Adenomas and Adenocarcinomas" "Ductal and Lobular Neoplasms" ...
>>   ..$ CMI-ASC              : chr "Soft Tissue Tumors and Sarcomas, NOS"
>>   ..$ BEATAML1.0-COHORT    : chr [1:6] "Leukemias, NOS" "Chronic Myeloproliferative Disorders" "Myeloid Leukemias" "Plasma Cell Tumors" ...
>>   ..$ CGCI-BLGSP           : chr "Mature B-Cell Lymphomas"
>>   ..$ BEATAML1.0-CRENOLANIB: chr "Myeloid Leukemias"
>>   ..$ CMI-MPC              : chr "Adenomas and Adenocarcinomas"
>>  $ name                  : chr [1:10] "Breast Invasive Carcinoma" "AACR Project GENIE - Contributed by Memorial Sloan Kettering Cancer Center" "AACR Project GENIE - Contributed by Vanderbilt-Ingram Cancer Center" "AACR Project GENIE - Contributed by Princess Margaret Cancer Centre" ...
>>  $ releasable            : logi [1:10] TRUE FALSE FALSE FALSE TRUE TRUE ...
>>  $ state                 : chr [1:10] "open" "open" "open" "open" ...
>>  $ released              : logi [1:10] TRUE TRUE TRUE TRUE TRUE TRUE ...
>>  - attr(*, "row.names")= int [1:10] 1 2 3 4 5 6 7 8 9 10
>>  - attr(*, "class")= chr [1:3] "GDCprojectsResults" "GDCResults" "list"
```

å¯ä»¥çœ‹åˆ°é»˜è®¤åªè¿”å› 10 æ¡è®°å½•ï¼Œæˆ‘ä»¬å¯ä»¥ä½¿ç”¨ `results()` çš„ `size` å’Œ `from` å‚æ•°æ¥æ”¹å˜éœ€è¦è¿”å›çš„æ•°é‡ï¼ˆsize è¡¨ç¤ºè¿”å›çš„è®°å½•æ•°ï¼Œfrom è¡¨ç¤ºä»å“ªä¸ªç´¢å¼•å¼€å§‹è¿”å›æ•°æ®ï¼‰ï¼›ä¹Ÿæœ‰ä¸€ä¸ªç®€ä¾¿çš„å‡½æ•° `results_all()` è¿”å›æ‰€æœ‰çš„è¯·æ±‚ç»“æœï¼Œå¯æƒ³è€ŒçŸ¥è¿™ç§æ–¹æ³•å¯èƒ½ä¼šå¯¼è‡´ä¸‹è½½çš„æ•°æ®éå¸¸å¤§ï¼ŒèŠ±è´¹çš„æ—¶é—´æ¯”è¾ƒä¹…ï¼Œå› æ­¤å°½é‡ä½¿ç”¨ `counts()` å’Œ `results()` æ¥è·å–æ•°æ®ã€‚

``` r
length(ids(presults))
>> [1] 10

presults = pquery %>% results_all()
length(ids(presults))
>> [1] 70

length(ids(presults)) == count(pquery)
>> [1] TRUE
```

ç»“æœæ˜¯ä»¥åˆ—è¡¨çš„å½¢å¼å­˜å‚¨çš„ï¼Œå¯ä»¥ä½¿ç”¨ `purrr`ï¼Œ`rlist`ï¼Œ`data.tree` ç­‰ R åŒ…æ¥æ“ä½œå¤æ‚åµŒå¥—çš„åˆ—è¡¨ï¼Œå¦å¤–ä¹Ÿå¯ä»¥ä½¿ç”¨ `listviewer` åŒ…æ¥äº¤äº’å¼çš„æŸ¥çœ‹åˆ—è¡¨çš„ç»“æ„ï¼ˆå…¶å®å’Œåœ¨ Rstudio ä¸­ç›´æ¥æ‰“å¼€å·®åˆ«ä¸å¤§ï¼‰ï¼š

``` r
listviewer::jsonedit(presults)
```

![](https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/image-20220108121517890.png)

### å­—æ®µå’Œå€¼

å‰é¢ä¹Ÿå·²ç»çœ‹åˆ°ä» GDC æ£€ç´¢å’Œè·å–æ•°æ®çš„ä¸­å¿ƒå°±æ˜¯æŒ‡å®šè¿”å›å“ªäº›å­—æ®µï¼Œæ ¹æ®å­—æ®µå’Œå€¼è¿›è¡Œç­›é€‰å¹¶è¿›è¡Œèšåˆç»Ÿè®¡ã€‚GenomicDataCommons æœ‰ä¸¤ä¸ªç®€å•çš„å‡½æ•°ï¼š`available_fields()` å’Œ `default_fields()`ï¼Œè¿™ä¸¤ä¸ªå‡½æ•°çš„å‚æ•°å¯ä»¥æ˜¯ä»£è¡¨è®¿é—®ç‚¹ï¼ˆendpointï¼‰åç§°çš„å­—ç¬¦ï¼ˆâ€œcasesâ€, â€œfilesâ€, â€œannotationsâ€, æˆ– â€œprojectsâ€ï¼‰æˆ–è€…æ˜¯ `GDCQuery`
å¯¹è±¡ï¼Œè¿”å›ç›¸åº”çš„å­—æ®µåç§°ï¼š

``` r
default_fields('files')
>>  [1] "access"                         "acl"                            "average_base_quality"          
>>  [4] "average_insert_size"            "average_read_length"            "channel"                       
>>  [7] "chip_id"                        "chip_position"                  "contamination"                 
>> [10] "contamination_error"            "created_datetime"               "data_category"                 
>> [13] "data_format"                    "data_type"                      "error_type"                    
>> [16] "experimental_strategy"          "file_autocomplete"              "file_id"                       
>> [19] "file_name"                      "file_size"                      "imaging_date"                  
>> [22] "magnification"                  "md5sum"                         "mean_coverage"                 
>> [25] "msi_score"                      "msi_status"                     "pairs_on_diff_chr"             
>> [28] "plate_name"                     "plate_well"                     "platform"                      
>> [31] "proportion_base_mismatch"       "proportion_coverage_10x"        "proportion_coverage_10X"       
>> [34] "proportion_coverage_30x"        "proportion_coverage_30X"        "proportion_reads_duplicated"   
>> [37] "proportion_reads_mapped"        "proportion_targets_no_coverage" "read_pair_number"              
>> [40] "revision"                       "stain_type"                     "state"                         
>> [43] "state_comment"                  "submitter_id"                   "tags"                          
>> [46] "total_reads"                    "tumor_ploidy"                   "tumor_purity"                  
>> [49] "type"                           "updated_datetime"
length(available_fields('files'))
>> [1] 981
head(available_fields('files'))
>> [1] "access"                      "acl"                         "analysis.analysis_id"        "analysis.analysis_type"     
>> [5] "analysis.created_datetime"   "analysis.input_files.access"
```

çŸ¥é“æœ‰å“ªäº›å­—æ®µåï¼Œæˆ‘ä»¬å°±å¯ä»¥é€šè¿‡ `select` åŠ¨è¯æ¥é€‰æ‹©éœ€è¦çš„å­—æ®µï¼›æ³¨æ„ï¼Œè¿™é‡Œçš„ `select` å’Œ `dplyr` çš„`select` å¹¶ä¸æ˜¯å®Œå…¨ä¸€æ ·çš„ï¼Œåè€…çš„ `select` åªèƒ½é€‰æ‹©å·²ç»å±•ç¤ºçš„å­—æ®µï¼Œè€Œè¿™é‡Œçš„ `select` å¯ä»¥æ‰€æœ‰çš„å¯é€‰å­—æ®µï¼ˆä¹Ÿå°±æ˜¯ default å’Œ available çš„åŒºåˆ«ï¼‰ï¼š

``` r
qcases = cases()
qcases$fields##Default fields 
>>  [1] "aliquot_ids"              "analyte_ids"              "case_autocomplete"        "case_id"                 
>>  [5] "consent_type"             "created_datetime"         "days_to_consent"          "days_to_lost_to_followup"
>>  [9] "diagnosis_ids"            "disease_type"             "index_date"               "lost_to_followup"        
>> [13] "portion_ids"              "primary_site"             "sample_ids"               "slide_ids"               
>> [17] "state"                    "submitter_aliquot_ids"    "submitter_analyte_ids"    "submitter_diagnosis_ids" 
>> [21] "submitter_id"             "submitter_portion_ids"    "submitter_sample_ids"     "submitter_slide_ids"     
>> [25] "updated_datetime"

default_fields(qcases)
>>  [1] "aliquot_ids"              "analyte_ids"              "case_autocomplete"        "case_id"                 
>>  [5] "consent_type"             "created_datetime"         "days_to_consent"          "days_to_lost_to_followup"
>>  [9] "diagnosis_ids"            "disease_type"             "index_date"               "lost_to_followup"        
>> [13] "portion_ids"              "primary_site"             "sample_ids"               "slide_ids"               
>> [17] "state"                    "submitter_aliquot_ids"    "submitter_analyte_ids"    "submitter_diagnosis_ids" 
>> [21] "submitter_id"             "submitter_portion_ids"    "submitter_sample_ids"     "submitter_slide_ids"     
>> [25] "updated_datetime"

##å¯ä»¥é€‰æ‹©default fieldä¸­æ²¡æœ‰çš„å­—æ®µ
qcases = cases() %>% GenomicDataCommons::select(available_fields('cases'))
head(qcases$fields)
>> [1] "case_id"                       "aliquot_ids"                   "analyte_ids"                  
>> [4] "annotations.annotation_id"     "annotations.case_id"           "annotations.case_submitter_id"
```

ç”±äºæ£€ç´¢å­—æ®µåœ¨ä¸‹è½½æ•°æ®è¿‡ç¨‹ä¸­æ˜¯ä¸€ä¸ªå¸¸ç”¨çš„æ“ä½œï¼Œå› æ­¤è¯¥åŒ…æä¾›äº†ä¸€äº›å‡½æ•°æ¥å¿«é€Ÿçš„æ‰¾åˆ°æƒ³è¦çš„å­—æ®µï¼Œå¦‚
`grep_fields()` å’Œ `field_picker()` ï¼ˆè²Œä¼¼ç°åœ¨çš„ç‰ˆæœ¬å·²ç»æŠŠ `field_picker()` å‡½æ•°ç»™åˆ äº†ï¼‰ã€‚

### èšåˆç»Ÿè®¡

ç»“åˆ `facet` å’Œ `aggregations()` å¯ä»¥å¯¹ä¸€ä¸ªæˆ–å¤šä¸ªå­—æ®µä¸­çš„å€¼è¿›è¡Œç»Ÿè®¡ï¼Œç±»ä¼¼ base R ä¸­çš„ `table` æ“ä½œï¼ˆä½†æ˜¯ä¸€æ¬¡åªèƒ½å¯¹ä¸€ä¸ªå­—æ®µæ“ä½œï¼Œä¸èƒ½è¿›è¡Œäº¤å‰æ“ä½œï¼‰ï¼Œè¿”å›ä¸€ä¸ªæ•°æ®æ¡†ï¼ˆtibblesï¼‰:

``` r
res = files() %>% facet(c('type','data_type')) %>% aggregations()
res$type
>>    doc_count                           key
>> 1     151568    annotated_somatic_mutation
>> 2      89689       simple_somatic_mutation
>> 3      87849                 aligned_reads
>> 4      66555               gene_expression
>> 5      58540           copy_number_segment
>> 6      45843          copy_number_estimate
>> 7      32268              mirna_expression
>> 8      30075                   slide_image
>> 9      25837        biospecimen_supplement
>> 10     14750          structural_variation
>> 11     13732        methylation_beta_value
>> 12     12962           clinical_supplement
>> 13      7906            protein_expression
>> 14      4410   aggregated_somatic_mutation
>> 15      4368       masked_somatic_mutation
>> 16      2746      masked_methylation_array
>> 17        54 secondary_expression_analysis
```

ä½¿ç”¨ `aggregations()` å¯ä»¥æ–¹ä¾¿çš„ç»Ÿè®¡ä¸€ä¸ªå­—æ®µä¸­æœ‰å“ªäº›å€¼ï¼Œæœ‰åˆ©äºæˆ‘ä»¬åç»­çš„é’ˆå¯¹å­—æ®µçš„ç­›é€‰ã€‚

### ç­›é€‰

GenomicDataCommons æä¾›äº†å’Œ dplyr çš„ filter åŒåä¸”ç±»ä¼¼çš„å‡½æ•° `filter()` ç”¨æ¥å¯¹ç‰¹å®šçš„å­—æ®µçš„å€¼è¿›è¡Œç­›é€‰ï¼Œæ¯”å¦‚åªæƒ³è¦ä¸Šé¢çš„ `type` ä¸­çš„ `gene_expression` æ•°æ®ï¼Œå°±å¯ä»¥è¿™æ ·æ¥ç­›é€‰ï¼š

``` r
qfiles = files() %>% filter( type == 'gene_expression')
str(get_filter(qfiles))
>> List of 2
>>  $ op     : 'scalar' chr "="
>>  $ content:List of 2
>>   ..$ field: chr "type"
>>   ..$ value: chr "gene_expression"
```

å¦‚æœæˆ‘ä»¬ç°åœ¨æƒ³è¦åŸºäºæŸä¸ª TCGA çš„æµ‹åºé¡¹ç›®æ¥æ„å»ºè¿‡æ»¤å™¨ï¼ˆæ¯”å¦‚ TCGA-OVï¼‰ï¼Œä½†æ˜¯ä¸çŸ¥é“å…·ä½“çš„å­—æ®µåç§°æ˜¯ä»€ä¹ˆï¼Œè¿™æ—¶å€™å°±å¯ä»¥ä½¿ç”¨ä¸Šé¢æåˆ°çš„ `grep_fields()` æ¥æŸ¥æ‰¾å¯èƒ½çš„å­—æ®µï¼š

``` r
grep_fields("files","project")
>>  [1] "cases.project.dbgap_accession_number"         "cases.project.disease_type"                  
>>  [3] "cases.project.intended_release_date"          "cases.project.name"                          
>>  [5] "cases.project.primary_site"                   "cases.project.program.dbgap_accession_number"
>>  [7] "cases.project.program.name"                   "cases.project.program.program_id"            
>>  [9] "cases.project.project_id"                     "cases.project.releasable"                    
>> [11] "cases.project.released"                       "cases.project.state"                         
>> [13] "cases.tissue_source_site.project"
```

çœ‹èµ·æ¥ `cases.project.project_id` æœ‰å¯èƒ½ç¬¦åˆæˆ‘ä»¬çš„è¦æ±‚ï¼Œæ¥ç€å†ç”¨ `facet` å’Œ `aggregations` æ¥æ£€æŸ¥è¯¥å­—æ®µä¸­æœ‰æ²¡æœ‰æˆ‘ä»¬æƒ³è¦çš„å€¼ï¼š

``` r
files() %>% 
    facet('cases.project.project_id') %>% 
    aggregations() %>% 
    head()
>> $cases.project.project_id
>>    doc_count                   key
>> 1      36134                 FM-AD
>> 2      34686             TCGA-BRCA
>> 3      45952               CPTAC-3
>> 4      18723             TCGA-LUAD
>> 5      36470             GENIE-MSK
>> 6      17717             TCGA-UCEC
>> 7      16694             TCGA-HNSC
>> 8      16776               TCGA-OV
>> 9      15826             TCGA-THCA
>> 10     29433         MMRF-COMMPASS
>> 11     16696             TCGA-LUSC
>> 12     16230              TCGA-LGG
>> 13     16733             TCGA-KIRC
>> 14     28464            GENIE-DFCI
>> 15     15648             TCGA-PRAD
>> 16     15701             TCGA-COAD
>> 17     13332              TCGA-GBM
>> 18     20772         TARGET-ALL-P2
>> 19     14025             TCGA-SKCM
>> 20     14096             TCGA-STAD
>> 21     12856             TCGA-BLCA
>> 22     11762             TCGA-LIHC
>> 23      9373             TCGA-CESC
>> 24      9353             TCGA-KIRP
>> 25      8228             TCGA-SARC
>> 26     14037             REBC-THYR
>> 27      8142            TARGET-AML
>> 28      5791             TCGA-PAAD
>> 29      5783             TCGA-ESCA
>> 30      5460             TCGA-PCPG
>> 31      5401             TCGA-READ
>> 32      9888               CPTAC-2
>> 33      5978             TCGA-TGCT
>> 34      8981     BEATAML1.0-COHORT
>> 35      5792            TARGET-NBL
>> 36      8167             HCMI-CMDC
>> 37      4814             TCGA-LAML
>> 38      5771         CGCI-HTMCP-CC
>> 39      3781             TCGA-THYM
>> 40      5941               CMI-MBC
>> 41      2782              TCGA-ACC
>> 42      2520             TCGA-KICH
>> 43      2677             TARGET-WT
>> 44      4805          NCICCR-DLBCL
>> 45      2580             TCGA-MESO
>> 46      2352              TCGA-UVM
>> 47      3113             TARGET-OS
>> 48      3982         TARGET-ALL-P3
>> 49      3857             GENIE-MDA
>> 50      3833            GENIE-VICC
>> 51      1813              TCGA-UCS
>> 52      3320             GENIE-JHU
>> 53      1456             TCGA-CHOL
>> 54      2632             GENIE-UHN
>> 55      1358             TCGA-DLBC
>> 56      2477            CGCI-BLGSP
>> 57      1049             TARGET-RT
>> 58      1038            GENIE-GRCC
>> 59       994            WCDT-MCRPC
>> 60       934               CMI-ASC
>> 61       801             GENIE-NKI
>> 62       798              OHSU-CNL
>> 63       703   ORGANOID-PANCREATIC
>> 64       570               CMI-MPC
>> 65       417           CTSP-DLBCL1
>> 66       339              TRIO-CRU
>> 67       222 BEATAML1.0-CRENOLANIB
>> 68       169           TARGET-CCSK
>> 69       133         TARGET-ALL-P1
>> 70        21        VAREPOP-APOLLO
```

è¿™ä¸ªå­—æ®µç¡®å®æ˜¯æœ‰ `TCGA-OV` çš„ï¼Œç„¶åå°±å¯ä»¥åŸºäºè¿™ä¸ªå­—æ®µçš„å€¼ä½¿ç”¨ `filter` è·å–æƒ³è¦çš„æ•°æ®ï¼š

``` r
qfiles = files() %>%
    filter( cases.project.project_id == 'TCGA-OV' & type == 'gene_expression')
str(get_filter(qfiles))
>> List of 2
>>  $ op     : 'scalar' chr "and"
>>  $ content:List of 2
>>   ..$ :List of 2
>>   .. ..$ op     : 'scalar' chr "="
>>   .. ..$ content:List of 2
>>   .. .. ..$ field: chr "cases.project.project_id"
>>   .. .. ..$ value: chr "TCGA-OV"
>>   ..$ :List of 2
>>   .. ..$ op     : 'scalar' chr "="
>>   .. ..$ content:List of 2
>>   .. .. ..$ field: chr "type"
>>   .. .. ..$ value: chr "gene_expression"

qfiles %>% count()
>> [1] 1137

##ä½¿ç”¨å¤šæ¬¡filterä¹Ÿæ˜¯å¯ä»¥çš„
qfiles2 = files() %>%
    filter( cases.project.project_id == 'TCGA-OV') %>% 
    filter( type == 'gene_expression') 
qfiles2 %>% count()
>> [1] 1137
```

æ£€ç´¢åˆ°æ•°æ®åå°±å¯ä»¥ä½¿ç”¨ `manidfest()`æ¥åŸºäºå½“å‰çš„è¯·æ±‚æ„å»ºéœ€è¦ä¸‹è½½æ•°æ®çš„å…ƒæ•°æ®ï¼š

``` r
manifest_df = qfiles %>% manifest()
>> Rows: 1137 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
head(manifest_df)
>> # A tibble: 6 Ã— 5
>>   id                                   filename                                             md5                      size state 
>>   <chr>                                <chr>                                                <chr>                   <dbl> <chr> 
>> 1 1c099182-703f-4aaf-b1d6-0f9be5094b9a fea3c4d0-7b9f-4279-989e-535aaefbdfeb.FPKM.txt.gz     a344c91e930a0217efa2bâ€¦ 506691 releaâ€¦
>> 2 02ab2f95-023b-492e-8dc9-20b4f52f209f b1d5c444-da0d-4360-bd45-31c94217adfc.FPKM-UQ.txt.gz  45d6b0fcc30607948a53bâ€¦ 547204 releaâ€¦
>> 3 30c9def0-c3b5-45b7-a427-020b735c836a ab8603dd-2f94-4c83-9927-455958be0007.FPKM.txt.gz     c82b63def1fcc6b9d3b52â€¦ 503463 releaâ€¦
>> 4 451c9c6d-f062-4ca2-8459-efed929bd305 9c6f2d76-7701-434e-a876-c04ab14cccea.htseq.counts.gz d296b0cb99d14b4b91383â€¦ 263970 releaâ€¦
>> 5 75c0968c-51aa-4f09-812f-7f85727d1e09 5b23f813-9970-485b-8577-ab2f3c029c26.FPKM-UQ.txt.gz  736a8c6706423a7c1ff40â€¦ 559501 releaâ€¦
>> 6 57d791ef-c2cc-4963-a618-05f92d5a4436 6aa7225b-3d10-4b19-a472-0adeb21c26cf.FPKM-UQ.txt.gz  8b44a1f944fc78b39e0f9â€¦ 536734 releaâ€¦
```

å¯ä»¥çœ‹åˆ°è¿”å›çš„ç»“æœä¸­æ—¢æœ‰ FPKM åˆæœ‰ FPKM-UQ å’Œ countsï¼Œæ¯”è¾ƒæ··ä¹±ï¼Œå¯ä»¥è¿›è¡Œè¿›ä¸€æ­¥çš„ç­›é€‰ï¼š

``` r
grep_fields("files","workflow")
>>  [1] "analysis.metadata.read_groups.read_group_qcs.workflow_end_datetime"  
>>  [2] "analysis.metadata.read_groups.read_group_qcs.workflow_link"          
>>  [3] "analysis.metadata.read_groups.read_group_qcs.workflow_start_datetime"
>>  [4] "analysis.metadata.read_groups.read_group_qcs.workflow_type"          
>>  [5] "analysis.metadata.read_groups.read_group_qcs.workflow_version"       
>>  [6] "analysis.workflow_end_datetime"                                      
>>  [7] "analysis.workflow_link"                                              
>>  [8] "analysis.workflow_start_datetime"                                    
>>  [9] "analysis.workflow_type"                                              
>> [10] "analysis.workflow_version"                                           
>> [11] "downstream_analyses.workflow_end_datetime"                           
>> [12] "downstream_analyses.workflow_link"                                   
>> [13] "downstream_analyses.workflow_start_datetime"                         
>> [14] "downstream_analyses.workflow_type"                                   
>> [15] "downstream_analyses.workflow_version"

##analysis.workflow_type
files() %>% 
    filter( cases.project.project_id == 'TCGA-OV') %>% 
    filter( type == 'gene_expression') %>% 
    facet('analysis.workflow_type') %>% 
    aggregations() %>% 
    head()
>> $analysis.workflow_type
>>    doc_count                                           key
>> 1       2400                                       DNAcopy
>> 2       1178                                        ASCAT2
>> 3       1051       BWA with Mark Duplicates and Cocleaning
>> 4        998                         BCGSC miRNA Profiling
>> 5        623                                      Liftover
>> 6        610                                       MuTect2
>> 7        610                            MuTect2 Annotation
>> 8        610                                 SomaticSniper
>> 9        610                      SomaticSniper Annotation
>> 10       610                                      VarScan2
>> 11       610                           VarScan2 Annotation
>> 12       606                                          MuSE
>> 13       606                               MuSE Annotation
>> 14       499                                       BWA-aln
>> 15       379                                HTSeq - Counts
>> 16       379                                  HTSeq - FPKM
>> 17       379                               HTSeq - FPKM-UQ
>> 18       379                                   STAR 2-Pass
>> 19         2          MuSE Variant Aggregation and Masking
>> 20         2       MuTect2 Variant Aggregation and Masking
>> 21         2 SomaticSniper Variant Aggregation and Masking
>> 22         2      VarScan2 Variant Aggregation and Masking
>> 23         1                    GISTIC - Copy Number Score
>> 24      3630                                      _missing

##HTSeq - Counts
qfiles = files() %>% filter( ~ cases.project.project_id == 'TCGA-OV' &
                            type == 'gene_expression' &
                            analysis.workflow_type == 'HTSeq - Counts')
manifest_df = qfiles %>% manifest()
>> Rows: 379 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
nrow(manifest_df)
>> [1] 379
```

## è®¾ç½® Token

GDC çš„æ•°æ®åˆ†ä¸º `controlled-access` å’Œ `open` ä¸¤ç§ç±»å‹ï¼›è¦ä¸‹è½½ `controlled-access` çš„æ•°æ®ï¼Œéœ€è¦ä¸€ä¸ª Token æ–‡ä»¶ï¼Œè¯¥æ–‡ä»¶å¯ä»¥ä» GDC å®˜ç½‘ä¸‹è½½ï¼ˆç™»å½•è´¦å·ä¹‹åï¼‰ï¼š

![](https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/%E5%B1%8F%E5%B9%95%E6%88%AA%E5%9B%BE%202022-01-08%20134754.png)

ç„¶åå¯ä»¥ç”¨ `gdc_token()` å‡½æ•°æ¥å¯¼å…¥ Tokenï¼Œè¯¥å‡½æ•°ä»¥ä¸‹é¢ä¸‰ç§æ–¹å¼ä¾æ¬¡æ¥æ‰¾ tokenï¼Œå¦‚æœæ²¡æœ‰æ‰¾åˆ°åˆ™æŠ¥é”™ï¼š

-   ä»¥ç¯å¢ƒå˜é‡ `GDC_TOKEN` ä¿å­˜å­—ç¬¦å½¢å¼çš„ token
-   ä»¥ç¯å¢ƒå˜é‡ `GDC_TOKEN_FILE` ä¿å­˜çš„ token æ–‡ä»¶åæ¥æ‰¾ token
-   åœ¨åœ¨å®¶ç›®å½•ä¸‹æ‰¾ä»¥æ–‡ä»¶å `.gdc_token` ä¿å­˜çš„ token æ–‡ä»¶

ä»¥ç¬¬ä¸‰ç§ä¸ºä¾‹ï¼Œå°†ä¸‹è½½çš„ token æ–‡ä»¶é‡å‘½åä¸º `.gdc_token`ï¼š

``` bash
cp data/gdc-user-token.2022-01-08T06_55_22.799Z.txt .gdc_token
```

``` r
token = gdc_token()
```

## æ•°æ®æ–‡ä»¶ä¸‹è½½

### é€šè¿‡ GDC API ä¸‹è½½æ•°æ®

`gdcdata` å¯ä»¥æ¥å—å«æœ‰ä¸€ä¸ªæˆ–å¤šä¸ªæ–‡ä»¶ UUID çš„å­—ç¬¦å‘é‡è¿›è¡Œæ–‡ä»¶çš„ä¸‹è½½ï¼Œç”Ÿæˆè¯¥å­—ç¬¦å‘é‡çš„ç®€å•æ–¹å¼å°±æ˜¯æŠŠä¸Šé¢è·å¾—çš„å…ƒæ•°æ®çš„ç¬¬ä¸€åˆ—æ‹¿å‡ºæ¥å°±è¡Œï¼š

``` r
fnames = gdcdata(manifest_df$id[1:2],progress=FALSE)
>> Rows: 2 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
```

å¦‚æœè¦ä¸‹è½½ control çš„æ•°æ®ï¼Œåˆ™éœ€è¦æä¾› tokenï¼ˆä¸Šé¢é€šè¿‡ `gdc_token` å¾—åˆ°çš„å­—ç¬¦ä¸²ï¼‰ï¼›å¦å¤–å¯ä»¥ä½¿ç”¨`BioCParallel` è¿›è¡Œå¤šçº¿ç¨‹ä¸‹è½½æ¥ç¼©çŸ­ä¸‹è½½çš„æ—¶é—´ã€‚

### å¤§æ•°æ®çš„ä¸‹è½½

å½“éœ€è¦ä¸‹è½½ä½“ç§¯å·¨å¤§çš„æ•°æ®ï¼Œæ¯”å¦‚æµ‹åºçš„ BAM æ–‡ä»¶æˆ–è€… è¾ƒå¤§çš„ VCF æ–‡ä»¶æ—¶ï¼Œå¯ä»¥ä½¿ç”¨ `client` çš„ä¸‹è½½æ–¹æ³•ï¼š

``` r
fnames = gdcdata(manifest_df$id[3:10], access_method = 'client')
>> Rows: 8 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
```

# æ€»ç»“

è¿™ä¸ªåŒ…ä¸‹è½½æ•°æ®é¦–å…ˆæ˜¯è¦æ„å»ºä¸€ä¸ª queryï¼ˆè¯·æ±‚ï¼‰ï¼Œæ¯”å¦‚ `files()` å’Œ `cases()`ï¼Œè¿™äº›å‡½æ•°è¿”å›ä¸€ä¸ª GDCQuery çš„ S3 å¯¹è±¡ï¼ŒåŒ…å« filter, facetsï¼Œå’Œå…¶ä»–çš„å‚æ•°ï¼›ç„¶åè¿›è¡Œç­›é€‰ï¼Œæœ‰å‡ ä¸ªé‡è¦çš„å‡½æ•°ï¼šfilter æ˜¯æ ¹æ® field çš„å€¼æ¥é€‰å–è®°å½•ï¼›select é€‰æ‹© fieldï¼›expand æ˜¯ field çš„èšåˆï¼Œæä¾›äº†ä¸€äº›ç›¸å…³çš„ field ç»„æˆçš„ groupï¼Œä¾¿äºæ£€ç´¢ï¼›facts æ˜¯ç»Ÿè®¡ common vlaueï¼Œç±»ä¼¼ tableï¼›æ¥ç€å°±å¯ä»¥é€šè¿‡ `manifest()` ä¸‹è½½å…ƒæ•°æ®ï¼Œé€šè¿‡å…ƒæ•°æ®ä¸‹è½½çœŸæ­£çš„æ•°æ®æ–‡ä»¶ï¼ˆ`gdcdata`ï¼‰

# æ¡ˆä¾‹ï¼šè®¡ç®— TCGA æ ·æœ¬çš„æµ‹åºæ·±åº¦

æµ‹åºæ·±åº¦çš„è®¡ç®—å…¬å¼ä¸ºï¼š

$$
D=\frac{L *N}{G}
$$
L æ˜¯ reads çš„é•¿åº¦ï¼ŒN æ˜¯ reads çš„æ•°ç›®ï¼ŒG æ˜¯æµ‹åºç›®æ ‡åŒºåŸŸçš„é•¿åº¦ï¼ˆæ¯”å¦‚å…¨å¤–æ˜¾å­ï¼Œå…¨åŸºå› ç»„ç­‰ï¼‰ï¼›å› æ­¤å¯¹äº TCGA çš„å…¨å¤–æ˜¾å­æµ‹åºï¼Œæˆ‘ä»¬éœ€è¦çŸ¥é“æ ·æœ¬æµ‹åºè¯»é•¿çš„é•¿åº¦ï¼Œreads çš„æ€»æ•°ç›®ï¼Œç›®æ ‡åŒºåŸŸå°±æ˜¯å¤–æ˜¾å­çš„é•¿åº¦ã€‚

é¦–å…ˆéœ€è¦å¾—åˆ°æ ·æœ¬æµ‹åºè¯»é•¿çš„é•¿åº¦ï¼Œä¹Ÿå°±æ˜¯ reads lengthï¼Œè¿™ä¸ªä¿¡æ¯å¯ä»¥ä»æµ‹åºçš„å…ƒæ•°æ®è·å–ï¼ˆä»£ç å‚è€ƒ[è¿™é‡Œ](https://support.bioconductor.org/p/107984/)ï¼‰ï¼š

``` r
library(GenomicDataCommons)

##ç­›é€‰æ•°æ®ï¼Œæ–¹ä¾¿æ¼”ç¤ºåªé€‰äº†æ ·æœ¬é‡æ¯”è¾ƒå°‘çš„ACC,å‰20ä¸ªæ ·æœ¬
q = files() %>% 
    filter( cases.project.project_id == 'TCGA-ACC') %>%
  filter(data_type == 'Aligned Reads' 
         & experimental_strategy == 'WXS' 
         & data_format == 'BAM') %>% 
  GenomicDataCommons::select('file_id') %>% 
  expand('analysis.metadata.read_groups') 

z = results(q,size=20)

library(dplyr)
>> 
>> Attaching package: 'dplyr'
>> The following objects are masked from 'package:GenomicDataCommons':
>> 
>>     count, filter, select
>> The following objects are masked from 'package:stats':
>> 
>>     filter, lag
>> The following objects are masked from 'package:base':
>> 
>>     intersect, setdiff, setequal, union
t <- z$analysis$metadata$read_groups
names(t) <- z$file_id
a <- t %>% bind_rows(.id = "file_ids") %>% as_tibble()
head(a)
>> # A tibble: 6 Ã— 19
>>   file_ids   sequencing_date experiment_name  target_capture_â€¦ submitter_id  target_capture_kit_â€¦ is_paired_end target_capture_â€¦
>>   <chr>      <chr>           <chr>            <chr>            <chr>         <chr>                <lgl>         <chr>           
>> 1 99818fa6-â€¦ 2013-09-12T08   TCGA-OR-A5KO-01â€¦ NimbleGen        efc0de29-4b5â€¦ http://www.nimblegeâ€¦ TRUE          06 465 668 001  
>> 2 a9f11831-â€¦ 2013-08-05T02   TCGA-OR-A5LD-10â€¦ NimbleGen        afc22b50-442â€¦ http://www.nimblegeâ€¦ TRUE          06 465 668 001  
>> 3 1a14bcc9-â€¦ 2013-08-12T02   TCGA-OR-A5JJ-10â€¦ NimbleGen        17fa8960-7a4â€¦ http://www.nimblegeâ€¦ TRUE          06 465 668 001  
>> 4 f3af7cc5-â€¦ 2013-08-12T03   TCGA-OR-A5JV-10â€¦ NimbleGen        b7c4c670-624â€¦ http://www.nimblegeâ€¦ TRUE          06 465 668 001  
>> 5 870c0429-â€¦ 2013-07-29T11   TCGA-OR-A5LP-01â€¦ NimbleGen        b99240a5-473â€¦ http://www.nimblegeâ€¦ TRUE          06 465 668 001  
>> 6 b1da0e48-â€¦ 2013-08-06T11   TCGA-OR-A5L2-01â€¦ NimbleGen        b99bd79e-aa3â€¦ http://www.nimblegeâ€¦ TRUE          06 465 668 001  
>> # â€¦ with 11 more variables: library_strategy <chr>, platform <chr>, created_datetime <chr>, updated_datetime <chr>,
>> #   read_group_name <chr>, library_name <chr>, target_capture_kit_name <chr>, sequencing_center <chr>, state <chr>,
>> #   read_length <int>, read_group_id <chr>

##æ•´ç†ç»“æœ
re <- a %>% 
  dplyr::select(experiment_name,read_length) %>% 
  mutate(sample=substr(experiment_name,1,16)) %>% 
  distinct(sample,read_length,.keep_all = T) %>% 
  dplyr::filter(grepl("TCGA",sample)) %>% 
  dplyr::filter(as.numeric(substr(sample,14,15)) <= 9)

table(re$read_length) ##read lengthéƒ½æ˜¯101
>> 
>> 101 
>>  13
```

ç„¶åéœ€è¦æ ¹æ® BAI æ–‡ä»¶å¾—åˆ° reads æ€»æ•°ç›®ï¼ˆæºä»£ç æ¥è‡ª[Coverage data for TCGA BAM files (biostars.org)](https://www.biostars.org/p/9472253/)ï¼‰ï¼š

``` r
#get BAM file manifest
manifest = GenomicDataCommons::files() %>%  
  GenomicDataCommons::filter( cases.project.project_id == 'TCGA-ACC') %>% 
  GenomicDataCommons::filter(experimental_strategy == "WXS" &
                               data_format == "BAM") %>%   
  GenomicDataCommons::manifest()
>> Rows: 184 Columns: 5
>> â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
>> Delimiter: "\t"
>> chr (4): id, filename, md5, state
>> dbl (1): size
>> 
>> â„¹ Use `spec()` to retrieve the full column specification for this data.
>> â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.

dt <- manifest %>% 
  mutate(tmp=gsub(".+[.]TCGA","TCGA",filename)) %>% 
  mutate(sample=substr(tmp,1,16)) %>% 
  dplyr::filter(!grepl("hg19",filename)) %>% 
  dplyr::filter(sample %in% re$sample)
```

ç”±äº BAM æ–‡ä»¶è¾ƒå¤§ï¼Œè€Œè®¡ç®— mapped çš„æ€» reads æ•°åªéœ€è¦ BAM çš„ç´¢å¼•æ–‡ä»¶ï¼ˆä¹Ÿå°±æ˜¯ BAI æ–‡ä»¶ï¼‰ï¼Œå› æ­¤æˆ‘ä»¬ä¸‹è½½ BAI æ–‡ä»¶å°±è¡Œï¼Œä½†æ˜¯ TCGA å¹¶æ²¡æœ‰ç›´æ¥æä¾› BAI æ–‡ä»¶çš„ UUIDï¼Œå› æ­¤æˆ‘ä»¬éœ€è¦è¿›è¡Œè¿›ä¸€æ­¥çš„å¤„ç†ã€‚æ ¹æ® GDC
å®˜æ–¹çš„[æ–‡æ¡£](https://gdc.cancer.gov/about-gdc/gdc-faqs)æ˜¾ç¤ºï¼Œåœ¨ç”¨ api ä¸‹è½½ BAM æ–‡ä»¶æ—¶åœ¨æœ«å°¾åŠ ä¸Š `?pretty=true&expand=index_files`å°±å¯ä»¥å¾—åˆ°ä¸€ä¸ªå«æœ‰ BAI çš„ UUID çš„ JSON æ–‡ä»¶ï¼š

![](https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/image-20220108143545648.png)

å› æ­¤æˆ‘ä»¬å¯ä»¥åˆ©ç”¨è¿™ä¸ªç‰¹å¾æ¥å¾—åˆ° BAI çš„ UUIDï¼š

``` r
res <- vector("list",length(unique(dt$id)))
for (i in 1:length(res)){
  con = curl::curl(paste0("https://api.gdc.cancer.gov/files/", dt$id[i], "?pretty=true&expand=index_files"))
  tbl = jsonlite::fromJSON(con)
  bai = data.frame(id = tbl$data$index_files$file_id,
                   filename = tbl$data$index_files$file_name,
                   md5 = tbl$data$index_files$md5sum,
                   size = tbl$data$index_files$file_size,
                   state = tbl$data$index_files$state)
  res[[i]] <- bai
  cat("complete",i,"\n")
}
>> complete 1 
>> complete 2 
>> complete 3 
>> complete 4 
>> complete 5 
>> complete 6 
>> complete 7 
>> complete 8 
>> complete 9 
>> complete 10 
>> complete 11 
>> complete 12 
>> complete 13
re <- bind_rows(res)
re
>>                                      id                                            filename                              md5
>> 1  a16a54a0-cd4a-43b7-8272-2565b0930bda TCGA-OR-A5KO-01A-11D-A29I-10_Illumina_gdc_realn.bai 8c6d16ebde6d101530ea3dcb7111fd9d
>> 2  327a4450-1c36-4677-9288-4a54a7d0d842 TCGA-OR-A5LP-01A-11D-A29I-10_Illumina_gdc_realn.bai 0214ad9219ec0df5d324682b291427a7
>> 3  4278fd34-4f28-4768-bb8e-c9388a702d39 TCGA-OR-A5L2-01A-11D-A30A-10_Illumina_gdc_realn.bai 271381422b85433078fc95b9f5018dbd
>> 4  bf2a9bea-c1d6-41c4-a11c-acfb24916c4c TCGA-PK-A5HC-01A-11D-A30A-10_Illumina_gdc_realn.bai 70c691fc38702fad1df7f10c97e0ce3c
>> 5  8047e7a5-4eaf-4518-9b8a-76c3c5bbd0c3 TCGA-OR-A5JR-01A-11D-A29I-10_Illumina_gdc_realn.bai 332d7af2eb8c669670f55ab688e27f45
>> 6  8e4e4082-9856-49c8-8d69-24ac7351014a TCGA-OR-A5LS-01A-11D-A29I-10_Illumina_gdc_realn.bai be312d99dfcd1efe4a98814c0a1a48ed
>> 7  907e80da-c646-414f-8d98-d5e3e7b4686f TCGA-OR-A5KZ-01A-11D-A29I-10_Illumina_gdc_realn.bai 574fec9fed318557d84435b043d947bf
>> 8  29daa003-4169-41cf-8452-83c06b6730e0 TCGA-OR-A5LA-01A-11D-A29I-10_Illumina_gdc_realn.bai c89f2e68dbf5bbf4cb1071ec7dbacfee
>> 9  7f59ebcb-a57d-458d-ba7b-9b77abb1f3f7 TCGA-OR-A5LK-01A-11D-A29I-10_Illumina_gdc_realn.bai 9330ad033f2eabb6909d3e834d80d14e
>> 10 00fb3422-6608-44ec-9afe-39e7bb0b52d4 TCGA-OR-A5L5-01A-11D-A29I-10_Illumina_gdc_realn.bai 67feea2ef796f4dc872f3d344b59880e
>> 11 c4595133-679a-4967-beb3-df0d9867ffa6 TCGA-OR-A5JJ-01A-11D-A29I-10_Illumina_gdc_realn.bai 303acd0b62ac94d46ce8ca7b46b0b874
>> 12 8fab8841-35f4-4c1e-a3f2-83a8d3fb4ab6 TCGA-OU-A5PI-01A-12D-A29I-10_Illumina_gdc_realn.bai e39845ba41c2ebe18f9bf9639a6962af
>> 13 950a115d-8102-47bc-aec2-5c3425a929fd TCGA-OR-A5JC-01A-11D-A29I-10_Illumina_gdc_realn.bai 88a3e25b655dcb9e3da954fd2b264192
>>       size    state
>> 1  6810304 released
>> 2  6797480 released
>> 3  7075944 released
>> 4  6882408 released
>> 5  6929096 released
>> 6  6898896 released
>> 7  8387160 released
>> 8  6714080 released
>> 9  6785888 released
>> 10 6834032 released
>> 11 6675856 released
>> 12 6776432 released
>> 13 6707696 released
# write.table(re,file = "~/data/TCGA_bai_manifest.txt",sep = "\t",row.names = F,quote = F)
```

ç„¶ååŸºäº BAI çš„ UUID æ¥ä¸‹è½½ BAI æ–‡ä»¶ï¼ˆè¿™é‡Œè¿˜æ˜¯ä½¿ç”¨äº† gdc-client æ¥ä¸‹è½½ï¼‰ï¼š

``` bash
../gdc-client download -m ../../TCGA_ACC_bai_manifest.txt -t ../gdc-user-token.2022-01-08T06_55_22.799Z.txt

ls > files
mkdir BAI
cat files | while read i;do cp ./$i/*.bai ./BAI/;done
ls BAI/
#TCGA-OR-A5JC-01A-11D-A29I-10_Illumina_gdc_realn.bai  TCGA-OR-A5LA-01A-11D-A29I-10_Illumina_gdc_realn.bai
#TCGA-OR-A5JJ-01A-11D-A29I-10_Illumina_gdc_realn.bai  TCGA-OR-A5LK-01A-11D-A29I-10_Illumina_gdc_realn.bai
#TCGA-OR-A5JR-01A-11D-A29I-10_Illumina_gdc_realn.bai  TCGA-OR-A5LP-01A-11D-A29I-10_Illumina_gdc_realn.bai
#TCGA-OR-A5KO-01A-11D-A29I-10_Illumina_gdc_realn.bai  TCGA-OR-A5LS-01A-11D-A29I-10_Illumina_gdc_realn.bai
#TCGA-OR-A5KZ-01A-11D-A29I-10_Illumina_gdc_realn.bai  TCGA-OU-A5PI-01A-12D-A29I-10_Illumina_gdc_realn.bai
#TCGA-OR-A5L2-01A-11D-A30A-10_Illumina_gdc_realn.bai  TCGA-PK-A5HC-01A-11D-A30A-10_Illumina_gdc_realn.bai
#TCGA-OR-A5L5-01A-11D-A29I-10_Illumina_gdc_realn.bai
```

å†åˆ©ç”¨ `samtools` çš„ idxstats åŠŸèƒ½ç»Ÿè®¡ readsï¼ˆè¿™é‡Œéšä¾¿æ‹¿ä¸€ä¸ª BAM æ–‡ä»¶å°±è¡Œï¼‰ï¼š

``` bash
ls *.bai > bai_files

#!/bin/bash
cat bai_files | while read i
do
  newname=`basename $i .bai`
  mv ../BAI/dummy.bam ../BAI/$newname.bam
  samtools idxstats ../BAI/$newname.bam > ./stat_files/$newname.txt
  mv ../BAI/$newname.bam ../BAI/dummy.bam
done
```

``` bash
head ~/data/TCGA_bai/test/stat_files/TCGA-OR-A5JC-01A-11D-A29I-10_Illumina_gdc_realn.txt
>> chr1 248956422   8655782 9484
>> chr10    133797422   6282801 6836
>> chr11    135086622   4863250 5439
>> chr11_KI270721v1_random  100316  3985679 4321
>> chr12    133275309   7341751 8237
>> chr13    114364328   5048809 5472
>> chr14    107043718   4231568 4745
>> chr14_GL000009v2_random  201709  3673268 4058
>> chr14_GL000225v1_random  211173  3236746 3564
>> chr14_KI270722v1_random  194050  4261146 4737
```

å¾—åˆ°çš„æ–‡ä»¶çš„ç¬¬ä¸‰åˆ—å°±æ˜¯å„ä¸ªæŸ“è‰²ä½“ mapped çš„æ€» readsï¼Œç„¶åå°±å¯ä»¥åˆ©ç”¨ä¸Šé¢çš„å…¬å¼è¿›è¡Œè®¡ç®—æµ‹åºæ·±åº¦äº†ï¼š

``` r
files <- list.files("~/data/TCGA_bai/test/stat_files/")
dt <- data.frame(file=files)
dt$samples <- gsub(".+[.]TCGA","TCGA",files) %>% 
  gsub("[.][0-9].+[_gdc_realn.txt]","",.) %>% 
  substr(.,1,16)
dt$depth <- NA
for (i in 1:nrow(dt)){
  a <- data.table::fread(paste0("~/data/TCGA_bai/test/stat_files/",dt$file[i]),data.table = F)
  depth <- ((sum(a$V3))/(38000000)) * 101
  dt$depth[i] <- depth
}
dt
>>                                                   file          samples    depth
>> 1  TCGA-OR-A5JC-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5JC-01A 253.7207
>> 2  TCGA-OR-A5JJ-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5JJ-01A 242.3565
>> 3  TCGA-OR-A5JR-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5JR-01A 290.4985
>> 4  TCGA-OR-A5KO-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5KO-01A 304.1783
>> 5  TCGA-OR-A5KZ-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5KZ-01A 463.0472
>> 6  TCGA-OR-A5L2-01A-11D-A30A-10_Illumina_gdc_realn.txt TCGA-OR-A5L2-01A 301.6314
>> 7  TCGA-OR-A5L5-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5L5-01A 283.6341
>> 8  TCGA-OR-A5LA-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5LA-01A 195.8037
>> 9  TCGA-OR-A5LK-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5LK-01A 267.7285
>> 10 TCGA-OR-A5LP-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5LP-01A 272.8540
>> 11 TCGA-OR-A5LS-01A-11D-A29I-10_Illumina_gdc_realn.txt TCGA-OR-A5LS-01A 313.4699
>> 12 TCGA-OU-A5PI-01A-12D-A29I-10_Illumina_gdc_realn.txt TCGA-OU-A5PI-01A 260.9008
>> 13 TCGA-PK-A5HC-01A-11D-A30A-10_Illumina_gdc_realn.txt TCGA-PK-A5HC-01A 329.6680
```
