

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=&#34;auto&#34;>



<head>
  <meta charset="UTF-8">
  <link rel="apple-touch-icon" sizes="76x76" href="/img/DNA.png">
  <link rel="icon" href="/img/DNA.png">
  <meta name="viewport"
        content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="description" content="">
  <meta name="author" content="wutao">
  <meta name="keywords" content="">
  
  <title>使用NeoPredPipe预测新抗原 - Hexo</title>

  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/css/bootstrap.min.css" />


  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/github-markdown-css@4.0.0/github-markdown.min.css" />
  <link  rel="stylesheet" href="/lib/hint/hint.min.css" />

  
    
    
      
      <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/highlight.js@10.6.0/styles/atelier-sulphurpool-dark.min.css" />
    
  

  
    <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.5.7/dist/jquery.fancybox.min.css" />
  



<!-- 主题依赖的图标库，不要自行修改 -->

<link rel="stylesheet" href="//at.alicdn.com/t/font_1749284_ba1fz6golrf.css">



<link rel="stylesheet" href="//at.alicdn.com/t/font_1736178_kmeydafke9r.css">


<link  rel="stylesheet" href="/css/main.css" />

<!-- 自定义样式保持在最底部 -->


  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    var CONFIG = {"hostname":"example.com","root":"/","version":"1.8.10","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"right","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"copy_btn":true,"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":true,"baidu":null,"google":null,"gtag":null,"tencent":{"sid":null,"cid":null},"woyaola":null,"cnzz":null,"leancloud":{"app_id":"A5lchRnum84Yumu5pVOSoVN8-MdYXbMMI","app_key":"runF3Mxw7648v64CISxiqMs8","server_url":null}}};
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
<meta name="generator" content="Hexo 5.4.0"></head>


<body>
  <header style="height: 70vh;">
    <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand"
       href="/">&nbsp;<strong>wutao's blog</strong>&nbsp;</a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/">
                <i class="iconfont icon-home-fill"></i>
                首页
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/">
                <i class="iconfont icon-archive-fill"></i>
                归档
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/">
                <i class="iconfont icon-category-fill"></i>
                分类
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/">
                <i class="iconfont icon-tags-fill"></i>
                标签
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/">
                <i class="iconfont icon-user-fill"></i>
                关于
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" data-toggle="modal" data-target="#modalSearch">&nbsp;<i
                class="iconfont icon-search"></i>&nbsp;</a>
          </li>
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self">&nbsp;<i
                class="iconfont icon-dark" id="color-toggle-icon"></i>&nbsp;</a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

    <div class="banner" id="banner" parallax=true
         style="background: url('/img/default.png') no-repeat center center;
           background-size: cover;">
      <div class="full-bg-img">
        <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
          <div class="page-header text-center fade-in-up">
            <span class="h2" id="subtitle" title="使用NeoPredPipe预测新抗原">
              
            </span>

            
              <div class="mt-3">
  
  
    <span class="post-meta">
      <i class="iconfont icon-date-fill" aria-hidden="true"></i>
      <time datetime="2021-03-25 17:03" pubdate>
        2021年3月25日 下午
      </time>
    </span>
  
</div>

<div class="mt-1">
  
    
    <span class="post-meta mr-2">
      <i class="iconfont icon-chart"></i>
      2.6k 字
    </span>
  

  
    
    <span class="post-meta mr-2">
      <i class="iconfont icon-clock-fill"></i>
      
      
      45
       分钟
    </span>
  

  
  
    
      <!-- LeanCloud 统计文章PV -->
      <span id="leancloud-page-views-container" class="post-meta" style="display: none">
        <i class="iconfont icon-eye" aria-hidden="true"></i>
        <span id="leancloud-page-views"></span> 次
      </span>
    
  
</div>

            
          </div>

          
        </div>
      </div>
    </div>
  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="d-none d-lg-block col-lg-2"></div>
    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div class="py-5" id="board">
          <article class="post-content mx-auto">
            <!-- SEO header -->
            <h1 style="display: none">使用NeoPredPipe预测新抗原</h1>
            
              <p class="note note-info">
                
                  本文最后更新于：2 个月前
                
              </p>
            
            <div class="markdown-body">
              <p><a target="_blank" rel="noopener" href="https://github.com/MathOnco/NeoPredPipe">NeoPredPipe</a>是一个可以从单区域或多区域测序得到的VCF文件来预测新抗原的流程工具,使用的注释软件为<code>ANNOVAR</code>;另外该软件在预测新抗原之后还有一个筛选步骤,这个筛选步骤是依据<a target="_blank" rel="noopener" href="https://www.nature.com/articles/nature24473">2017 Nature</a>提出的<code>Neoantigen recognition potential</code>来进行的</p>
<h2 id="安装ANNOVAR"><a href="#安装ANNOVAR" class="headerlink" title="安装ANNOVAR"></a>安装ANNOVAR</h2><p>在<a target="_blank" rel="noopener" href="https://annovar.openbioinformatics.org/en/latest/user-guide/download/">官网</a> 下载ANNOVAR软件(需要填申请表,下载链接会发到邮箱), ANNOVAR是用perl写的,所以在安装之前需要先下载安装perl</p>
<p>下载解压后将路径添加到PATH中,向<code>~/.bashrc</code>中添加：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-built_in">export</span> PATH=<span class="hljs-string">&quot;/public/slst/home/wutao2/software/annovar:<span class="hljs-variable">$PATH</span>&quot;</span><br></code></pre></td></tr></table></figure>

<p>使用<code>source ~/.bashrc </code>激活PATH</p>
<p>下载注释所需的参考文件,这里下载的是<code>hg38的refgene</code>：</p>
<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/image-20210325085447536.png" srcset="/img/loading.gif" lazyload></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs bash">perl annotate_variation.pl --downdb --webfrom annovar --buildver hg38 refGene ~/software/annovar/humandb/<br><span class="hljs-comment"># -downdb表明该命令的用途是下载数据库</span><br><span class="hljs-comment"># -buildver指定基因组版本</span><br><span class="hljs-comment"># -webform annovar 从annovar提供的镜像下载</span><br><span class="hljs-comment"># refGene代表的是下载的数据库的名字</span><br><span class="hljs-comment">#~/software/annovar/humandb/表示数据库存储的路径</span><br></code></pre></td></tr></table></figure>

<p>下载的文件为：<code>hg38_refGeneMrna.fa</code>,<code>hg38_refGene.txt</code>和<code>hg38_refGeneVersion.txt</code></p>
<h2 id="下载PeptideMatch和参考多肽序列"><a href="#下载PeptideMatch和参考多肽序列" class="headerlink" title="下载PeptideMatch和参考多肽序列"></a>下载PeptideMatch和参考多肽序列</h2><p> <code>PeptideMatch</code>可以用来将得到的新抗原肽与参考肽序列进行比对,进而检查预测的<code>neoantigen</code>是不是”新”的</p>
<p><code>PeptideMatch</code>的下载地址在<a target="_blank" rel="noopener" href="https://research.bioinformatics.udel.edu/peptidematch/commandlinetool.jsp">here</a>(需要安装<code>java</code>)</p>
<p>另外还需要下载<code>fasta </code>格式的参考蛋白序列,数据在<a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/reference_proteomes/">here</a>,可以看到人的数据为<a href="ftp://ftp.ebi.ac.uk/pub/databases/reference_proteomes/QfO/Eukaryota/UP000005640_9606.fasta.gz">UP000005640 9606 HUMAN Homo sapiens</a> 从FTP下载得到<code>UP000005640_9606.fasta</code>文件,然后利用<code>PeptideMatch</code>处理得到<code>index</code>:</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs shell">java -jar ~/software/PeptideMatchCMD_1.0.jar -a index -d UP000005640_9606.fasta -i UP000005640_9606_index<br><br>Command line options: -a index -d UP000005640_9606.fasta -i UP000005640_9606_index <br>Indexing to directory &quot;/slst/home/wutao2/protein_database/UP000005640_9606_index&quot; ...<br>Indexing &quot;UP000005640_9606.fasta&quot; ...<br>Indexing &quot;UP000005640_9606.fasta&quot; finished<br>Time used: 00 hours, 00 mins, 24.869 seconds<br></code></pre></td></tr></table></figure>

<h2 id="安装和配置NeoPredPipe"><a href="#安装和配置NeoPredPipe" class="headerlink" title="安装和配置NeoPredPipe"></a>安装和配置NeoPredPipe</h2><p>该工具使用的是<code>python2.7</code>,所以需要先创建一个<code>python2</code>的<code>conda</code>环境：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs bash">mamba create -n python27 python=2.7.13<br><br><span class="hljs-comment">###安装依赖</span><br>pip config <span class="hljs-built_in">set</span> global.index-url https://pypi.tuna.tsinghua.edu.cn/simple<span class="hljs-comment">##设置镜像</span><br>python -m pip install biopython==1.70<br></code></pre></td></tr></table></figure>

<p>然后下载安装该软件</p>
<p>首先克隆<code>github</code>仓库：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs bash">git <span class="hljs-built_in">clone</span> https://github.com/MathOnco/NeoPredPipe.git<br></code></pre></td></tr></table></figure>

<p>然后需要配置<code>usr_path.ini</code>文件,将原始的文件内容进行修改为：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><code class="hljs bash">[annovar]<br>convert2annovar = /public/slst/home/wutao2/software/annovar/convert2annovar.pl<br>annotatevariation = /public/slst/home/wutao2/software/annovar/annotate_variation.pl<br>coding_change =/public/slst/home/wutao2/software/annovar/coding_change.pl<br>gene_table = /public/slst/home/wutao2/software/annovar/humandb/hg38_refGene.txt<br>gene_fasta =/public/slst/home/wutao2/software/annovar/humandb/hg38_refGeneMrna.fa<br>humandb =/public/slst/home/wutao2/software/annovar/humandb/<br>[netMHCpan]<br>netMHCpan = /public/slst/home/wutao2/software/netMHCpan-4.1/netMHCpan<br>[PeptideMatch]<br>peptidematch_jar = /public/slst/home/wutao2/software/PeptideMatchCMD_1.0.jar<br>reference_index = /public/slst/home/wutao2/protein_database/UP000005640_9606_index/<br>[blast]<br>blastp =/public/slst/home/wutao2/software/ncbi-blast-2.11.0+/bin/blastp<br></code></pre></td></tr></table></figure>

<p>测试安装是否成功:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br></pre></td><td class="code"><pre><code class="hljs bash">python ./NeoPredPipe.py --<span class="hljs-built_in">help</span><br>usage: NeoPredPipe.py [-h] [-E EPITOPES [EPITOPES ...]] [-l] [-d] [-r] [-p]<br>                      [--manualproc] [--EL] [-I VCFDIR] [-H HLAFILE]<br>                      [-o OUTPUTDIR] [-n OUTNAME] [-pp]<br>                      [-c COLREGIONS [COLREGIONS ...]] [-a] [-m]<br>                      [-x EXPRESSION] [--expmulti] [-t]<br><br>optional arguments:<br>  -h, --<span class="hljs-built_in">help</span>            show this <span class="hljs-built_in">help</span> message and <span class="hljs-built_in">exit</span><br>  -E EPITOPES [EPITOPES ...], --epitopes EPITOPES [EPITOPES ...]<br>                        Epitope lengths <span class="hljs-keyword">for</span> predictions. Default: 8 9 10<br>  -l                    Specifies whether to delete the ANNOVAR <span class="hljs-built_in">log</span> file.<br>                        Default: True. Note: Use <span class="hljs-keyword">for</span> debugging.<br>  -d                    Specifies whether to delete intermediate files created<br>                        by program. Default: True. Note: Set flag to resume<br>                        job.<br>  -r, --cleanrun        Specify this alone with no other options to clean-up a<br>                        run. Be careful that you mean to <span class="hljs-keyword">do</span> this!!<br>  -p, --preponly        Prep files only without running neoantigen<br>                        predictions. The prediction step takes the most time.<br>  --manualproc          Process vcf files into annovar-input format manually,<br>                        to avoid issues from non <span class="hljs-string">&#x27;genotype-calling&#x27;</span> formats.<br>  --EL                  Flag to perform netMHCpan predictions with Eluted<br>                        Ligand option (without the -BA flag). Please note that<br>                        the output will NOT be compatible with downstream<br>                        Recognition Potential analysis. Default=False (BA<br>                        predictions)<br><br>Required arguments:<br>  -I VCFDIR             Input vcf file directory location. Example: -I<br>                        ./Example/input_vcfs/<br>  -H HLAFILE            HLA file <span class="hljs-keyword">for</span> vcf patient samples OR directory with<br>                        patient-specific directories from running POLYSOLVER<br>                        (see Readme).<br>  -o OUTPUTDIR          Output Directory Path<br>  -n OUTNAME            Name of the output file <span class="hljs-keyword">for</span> neoantigen predictions<br><br>Post Processing Options:<br>  -pp                   Flag to perform post processing. Default=True.<br>  -c COLREGIONS [COLREGIONS ...]<br>                        Columns of regions within vcf that are not normal<br>                        within a multiregion vcf file after the format field.<br>                        Example: 0 is normal <span class="hljs-keyword">in</span> <span class="hljs-built_in">test</span> samples, tumor are the<br>                        other columns. Program can handle different number of<br>                        regions per vcf file.<br>  -a                    Flag to not filter neoantigen predictions and keep all<br>                        regardless of prediction value.<br>  -m                    Specifies whether to perform check <span class="hljs-keyword">if</span> predicted<br>                        epitopes match any normal peptide. If <span class="hljs-built_in">set</span> to True,<br>                        output is added as a column to neoantigens file.<br>                        Requires PeptideMatch specified <span class="hljs-keyword">in</span> usr_paths.ini.<br>                        Default=False<br>  -x EXPRESSION, --expression EXPRESSION<br>                        RNAseq expression quantification file(s), <span class="hljs-keyword">if</span><br>                        specified, expression information is added to output<br>                        tables.<br>  --expmulti            Flag to specify <span class="hljs-keyword">if</span> expression file(s) has information<br>                        on multiple regions <span class="hljs-keyword">in</span> multiple columns.<br>                        Default=False.<br>  -t                    Flag to turn off a neoantigen burden summary table.<br>                        Default=True.<br></code></pre></td></tr></table></figure>

<p>该软件的输入文件有：</p>
<ul>
<li><p>VCF文件：可以是单区域测序也可以是多区域测序</p>
</li>
<li><p>hla文件：hla文件的格式如下</p>
<table>
<thead>
<tr>
<th>Patient</th>
<th>HLA-A_1</th>
<th>HLA-A_2</th>
<th>HLA-B_1</th>
<th>HLA-B_2</th>
<th>HLA-C_1</th>
<th>HLA-C_2</th>
</tr>
</thead>
<tbody><tr>
<td>test1</td>
<td>hla_a_31_01_02</td>
<td>hla_a_02_01_80</td>
<td>hla_b_40_01_02</td>
<td>hla_b_50_01_01</td>
<td>hla_c_03_04_20</td>
<td>hla_c_06_02_01_02</td>
</tr>
<tr>
<td>test2</td>
<td>hla_a_01_01_01_01</td>
<td>NA</td>
<td>hla_b_07_02_01</td>
<td>NA</td>
<td>hla_c_01_02_01</td>
<td>NA</td>
</tr>
</tbody></table>
<p>  Patient名称要和vcf文件的名称相匹配；制表符分割；可以不要列名,但是顺序要匹配；当两个位点预测的HLA是一样的时候(A1和A2,B1和B2,C1和C2),需要用NA代替</p>
</li>
<li><p>表达文件：在-x参数后指定,制表符分割,第一列是gene id第二列是表达值;支持的id有：Ensembl gene ID, Ensembl transcript ID, RefSeq transcript ID, UCSC transcript ID</p>
</li>
</ul>
<p>使用测试数据进行测试(该软件提供的测试数据的参考基因组是<code>hg19</code>,因此将上面配置文件中的<code>gene_table</code>和<code>gene_fasta</code>改成<code>hg19</code>的)：</p>
<figure class="highlight sh"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs sh">NeoPredPipe.py --preponly -I ~/software/NeoPredPipe/Example/input_vcfs -H   ~/software/NeoPredPipe/Example/HLAtypes/hlatypes.txt -o ./<span class="hljs-built_in">test</span>/ -n TestRun -c 1 2 -E 8 9 10<br><br>INFO: Annovar reference files of build hg19 were given, using this build <span class="hljs-keyword">for</span> all analysis.<br>INFO: Begin.<br>INFO: Running convert2annovar.py on /public/slst/home/wutao2/software/NeoPredPipe/Example/input_vcfs/test1.vcf<br>INFO: ANNOVAR VCF Conversion Process complete /public/slst/home/wutao2/software/NeoPredPipe/Example/input_vcfs/test1.vcf<br>INFO: Running annotate_variation.pl on ./<span class="hljs-built_in">test</span>/avready/test1.avinput<br>INFO: ANNOVAR annotation Process complete <span class="hljs-keyword">for</span> ./<span class="hljs-built_in">test</span>/avready/test1.avinput<br>INFO: Running coding_change.pl on ./<span class="hljs-built_in">test</span>/avannotated/test1.avannotated.exonic_variant_function<br>INFO: Coding predictions complete <span class="hljs-keyword">for</span> ./<span class="hljs-built_in">test</span>/avannotated/test1.avannotated.exonic_variant_function<br>INFO: Input files prepared and completed <span class="hljs-keyword">for</span> test1<br>INFO: Running convert2annovar.py on /public/slst/home/wutao2/software/NeoPredPipe/Example/input_vcfs/test2.vcf<br>INFO: ANNOVAR VCF Conversion Process complete /public/slst/home/wutao2/software/NeoPredPipe/Example/input_vcfs/test2.vcf<br>INFO: Running annotate_variation.pl on ./<span class="hljs-built_in">test</span>/avready/test2.avinput<br>INFO: ANNOVAR annotation Process complete <span class="hljs-keyword">for</span> ./<span class="hljs-built_in">test</span>/avready/test2.avinput<br>INFO: Running coding_change.pl on ./<span class="hljs-built_in">test</span>/avannotated/test2.avannotated.exonic_variant_function<br>INFO: Coding predictions complete <span class="hljs-keyword">for</span> ./<span class="hljs-built_in">test</span>/avannotated/test2.avannotated.exonic_variant_function<br>INFO: Input files prepared and completed <span class="hljs-keyword">for</span> test2<br>INFO: Complete.<br>INFO: Preprocessed intermediary files are <span class="hljs-keyword">in</span> avready, avannotated and fastaFiles. If you wish to perform epitope prediction, run the pipeline again without the --preponly flag, intermediary files will be automatically detected.<br></code></pre></td></tr></table></figure>

<p>这一步是准备输入文件的,也就是运行<code>ANNOVAR</code>将变异进行注释得到多肽序列(<code>convert2annovar.py</code>,<code>annotate_variation.pl</code>,和<code>coding_change.pl </code>)</p>
<p>也可以直接预测得到结果：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><code class="hljs bash">NeoPredPipe.py -I ~/software/NeoPredPipe/Example/input_vcfs -H ~/software/NeoPredPipe/Example/HLAtypes/hlatypes.txt -o ./test_results/ -n TestRun -c 1 2 -E 8 9 10<br><br>INFO: Annovar reference files of build hg19 were given, using this build <span class="hljs-keyword">for</span> all analysis.<br>INFO: Begin.<br>INFO: Running convert2annovar.py on /public/slst/home/wutao2/software/NeoPredPipe/Example/input_vcfs/test1.vcf<br>INFO: ANNOVAR VCF Conversion Process complete /public/slst/home/wutao2/software/NeoPredPipe/Example/input_vcfs/test1.vcf<br>INFO: Running annotate_variation.pl on ./test_results/avready/test1.avinput<br>INFO: ANNOVAR annotation Process complete <span class="hljs-keyword">for</span> ./test_results/avready/test1.avinput<br>INFO: Running coding_change.pl on ./test_results/avannotated/test1.avannotated.exonic_variant_function<br>INFO: Coding predictions complete <span class="hljs-keyword">for</span> ./test_results/avannotated/test1.avannotated.exonic_variant_function<br>INFO: Predicting neoantigens <span class="hljs-keyword">for</span> test1<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test1 on epitopes of length 9.Indels<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test1 on epitopes of length 9<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test1 on epitopes of length 8<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test1 on epitopes of length 8.Indels<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test1 on epitopes of length 10<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test1 on epitopes of length 10.Indels<br>INFO: Predictions complete <span class="hljs-keyword">for</span> test1 on epitopes of length 10.Indels<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test1<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test1<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test1<br>INFO: Object size of neoantigens: 48472 Kb<br>Processing genotype information according to A (list of alleles) field.<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test1<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test1<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test1<br>INFO: Object size of neoantigens: 3768 Kb<br>Processing genotype information according to A (list of alleles) field.<br>INFO: Running convert2annovar.py on /public/slst/home/wutao2/software/NeoPredPipe/Example/input_vcfs/test2.vcf<br>INFO: ANNOVAR VCF Conversion Process complete /public/slst/home/wutao2/software/NeoPredPipe/Example/input_vcfs/test2.vcf<br>INFO: Running annotate_variation.pl on ./test_results/avready/test2.avinput<br>INFO: ANNOVAR annotation Process complete <span class="hljs-keyword">for</span> ./test_results/avready/test2.avinput<br>INFO: Running coding_change.pl on ./test_results/avannotated/test2.avannotated.exonic_variant_function<br>INFO: Coding predictions complete <span class="hljs-keyword">for</span> ./test_results/avannotated/test2.avannotated.exonic_variant_function<br>INFO: Predicting neoantigens <span class="hljs-keyword">for</span> test2<br>INFO: Skipping Sample! No peptides to predict <span class="hljs-keyword">for</span> test2<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test2 on epitopes of length 9<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test2 on epitopes of length 8<br>INFO: Skipping Sample! No peptides to predict <span class="hljs-keyword">for</span> test2<br>INFO: Running Epitope Predictions <span class="hljs-keyword">for</span> test2 on epitopes of length 10<br>INFO: Skipping Sample! No peptides to predict <span class="hljs-keyword">for</span> test2<br>INFO: Predictions complete <span class="hljs-keyword">for</span> test2 on epitopes of length 10.Indels<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test2<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test2<br>INFO: Digesting neoantigens <span class="hljs-keyword">for</span> test2<br>INFO: Object size of neoantigens: 26744 Kb<br>Processing genotype information according to A (list of alleles) field.<br>INFO: Summary Tables Complete.<br>INFO: Summary Tables Complete.<br>INFO: Complete<br></code></pre></td></tr></table></figure>

<p>输出文件是没有表头的,每列的信息为：</p>
<blockquote>
<p>The primary output file of neoantigens has the following format, separated by tabulators (columns 12-26 are taken from <a target="_blank" rel="noopener" href="http://www.cbs.dtu.dk/services/NetMHCpan/output.php">here</a>):</p>
<ul>
<li><strong>Sample</strong>: vcf filename/patient identifier</li>
<li><strong>R1</strong>: Region 1 of a multiregion sample, binary for presence (1) or absence (0), regions above the number of regions in the sample (for varying number of biopsies) are indicated by -1. Can be <em>n</em> numbers of regions. <em>Only present in multiregion samples</em>.</li>
<li><strong>R2</strong>: Region 2 of a multiregion sample, binary for presence (1) or absence (0), regions above the number of regions in the sample (for varying number of biopsies) are indicated by -1. Can be <em>n</em> numbers of regions. <em>Only present in multiregion samples</em>.</li>
<li><strong>R3</strong>: Region 3 of a multiregion sample, binary for presence (1) or absence (0), regions above the number of regions in the sample (for varying number of biopsies) are indicated by -1. Can be <em>n</em> numbers of regions. <em>Only present in multiregion samples</em>.</li>
<li><strong>Line</strong>: Line number from the *.avready file (same as the vcf) to identify mutation yielding corresponding neoantigen.</li>
<li><strong>chr</strong>: Chromosome of mutation</li>
<li><strong>allelepos</strong>: Position of the mutation</li>
<li><strong>ref</strong>: Reference base at the position</li>
<li><strong>alt</strong>: Alternative base at the location</li>
<li><strong>GeneName:RefID</strong>: Gene name and RefSeq ID separated by a colon. Multiple genes/RefSeq IDs separated by a comma.</li>
<li><strong>Expression</strong>: Expression value of the gene. Expression values for multiple regions (<em>if using the -expmulti flag</em>) are comma-separated. NA for genes that are not found in the corresponding expression file, or for samples without expression information. <em>Only present if the -x flag is used</em>.</li>
<li><strong>pos</strong>: Residue number (starting from 0)</li>
<li><strong>hla</strong>: Molecule/allele name</li>
<li><strong>peptide</strong>: Amino acid sequence of the potential ligand</li>
<li><strong>core</strong>: The minimal 9 amino acid binding core directly in contact with the MHC</li>
<li><strong>Of</strong>: The starting position of the Core within the Peptide (if &gt; 0, the method predicts a N-terminal protrusion)</li>
<li><strong>Gp</strong>: Position of the deletion, if any.</li>
<li><strong>Gl</strong>: Length of the deletion.</li>
<li><strong>Ip</strong>: Position of the insertions, if any.</li>
<li><strong>Il</strong>: Length of the insertion.</li>
<li><strong>Icore</strong>: Interaction core. This is the sequence of the binding core including eventual insertions of deletions.</li>
<li><strong>Identity</strong>: Protein identifier, i.e. the name of the Fasta entry.</li>
<li><strong>Score</strong>: The raw prediction score</li>
<li><strong>Binding Affinity</strong>: Predicted binding affinity in nanoMolar units.</li>
<li><strong>Rank</strong>: Rank of the predicted affinity compared to a set of random natural peptides. This measure is not affected by inherent bias of certain molecules towards higher or lower mean predicted affinities. Strong binders are defined as having %rank&lt;0.5, and weak binders with %rank&lt;2. We advise to select candidate binders based on %Rank rather than nM Affinity</li>
<li><strong>Candidate</strong>: Symbol (&lt;=) used to denote a Strong or Week Binder in BindLevel</li>
<li><strong>BindLevel</strong>: (SB: strong binder, WB: weak binder). The peptide will be identified as a strong binder if the % Rank is below the specified threshold for the strong binders, by default 0.5%. The peptide will be identified as a weak binder if the % Rank is above the threshold of the strong binders but below the specified threshold for the weak binders, by default 2%.</li>
<li><strong>Novelty</strong>: Binary value for indicating if the epitope is novel (1) or exists in the reference proteome (0). <em>Only present if -m flag is set to perform peptide matching in postprocessing</em>.</li>
</ul>
</blockquote>
<h2 id="Recognition-Potential"><a href="#Recognition-Potential" class="headerlink" title="Recognition Potential"></a>Recognition Potential</h2><p>在预测新抗原之后,该工具还可以根据2017 nature文章的方法计算neoantigen recognition potential</p>
<h2 id="注意事项"><a href="#注意事项" class="headerlink" title="注意事项"></a>注意事项</h2><ul>
<li><p><code>-c</code>参数的含义：</p>
<p><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/image-20210324145740028.png" srcset="/img/loading.gif" lazyload><code>-c</code>参数表示从<code>FORMAT</code>列往后第几列是需要关注的(一般是tumor列);在单区域测序的样本中可以不指定</p>
</li>
<li><p>如果有表达数据,表达数据的<code>gene id</code>如果是<code>Ensembl gene ID</code>(也就是<code>ENSG</code>开头),要去掉版本号(小数点后的数字),因为该软件进行ID转化时使用的文件<a target="_blank" rel="noopener" href="https://github.com/MathOnco/NeoPredPipe/blob/master/mart_table_hg38_unique.txt">mart_table_hg38_unique.txt</a> 时不带版本号的</p>
</li>
<li><p>HLA的格式需要是像HLA-A03:01这样的</p>
</li>
<li><p>输出的列：注意netMHCpan的版本<br><img src="https://picgo-wutao.oss-cn-shanghai.aliyuncs.com/img/image-20210326124059938.png" srcset="/img/loading.gif" lazyload></p>
</li>
</ul>

            </div>
            <hr>
            <div>
              <div class="post-metas mb-3">
                
                  <div class="post-meta mr-3">
                    <i class="iconfont icon-category"></i>
                    
                      <a class="hover-with-bg" href="/categories/skills/">skills</a>
                    
                  </div>
                
                
                  <div class="post-meta">
                    <i class="iconfont icon-tags"></i>
                    
                      <a class="hover-with-bg" href="/tags/bioinformatics/">bioinformatics</a>
                    
                  </div>
                
              </div>
              
                <p class="note note-warning">
                  
                    本博客所有文章除特别声明外，均采用 <a target="_blank" href="https://creativecommons.org/licenses/by-sa/4.0/deed.zh" rel="nofollow noopener noopener">CC BY-SA 4.0 协议</a> ，转载请注明出处！
                  
                </p>
              
              
                <div class="post-prevnext">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/2021/03/28/lagrange_duality/">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">拉格朗日对偶性</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/2021/03/19/snakemake/">
                        <span class="hidden-mobile">使用Snakemake搭建流程</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
              <!-- Comments -->
              <article class="comments" id="comments" lazyload>
                
                  
                
                
  <div id="waline"></div>
  <script type="text/javascript">
    Fluid.utils.loadComments('#waline', function() {
      Fluid.utils.createScript('https://cdn.jsdelivr.net/npm/@waline/client@0.14.8/dist/Waline.min.js', function () {
        new Waline({
          el: "#waline",
          placeholder: "说点什么",
          path: window.location.pathname,
          avatar: "retro",
          meta: ["nick","mail","link"],
          pageSize: "5",
          lang: "zh-CN",
          highlight: true,
          serverURL: "https://comments-flax.vercel.app",
          avatarCDN: "",
          avatarForce: false,
          requiredFields: [],
          emojiCDN: "https://cdn.jsdelivr.net/gh/walinejs/emojis@1.0.0/bilibili",
          emojiMaps: {},
          anonymous: null,
        });
      });
    });
  </script>
  <noscript>Please enable JavaScript to view the comments</noscript>


              </article>
            
          </article>
        </div>
      </div>
    </div>
    
      <div class="d-none d-lg-block col-lg-2 toc-container" id="toc-ctn">
        <div id="toc">
  <p class="toc-header"><i class="iconfont icon-list"></i>&nbsp;目录</p>
  <div class="toc-body" id="toc-body"></div>
</div>

      </div>
    
  </div>
</div>

<!-- Custom -->


    

    
      <a id="scroll-top-button" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v"
                 for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>
    

    
  </main>

  <footer class="text-center mt-5 py-3">
  <div class="footer-content">
     <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
  </div>
  
  <div class="statistics">
    
    

    
      
        <!-- LeanCloud 统计PV -->
        <span id="leancloud-site-pv-container" style="display: none">
            总访问量 
            <span id="leancloud-site-pv"></span>
             次
          </span>
      
      
        <!-- LeanCloud 统计UV -->
        <span id="leancloud-site-uv-container" style="display: none">
            总访客数 
            <span id="leancloud-site-uv"></span>
             人
          </span>
      

    
  </div>


  

  
</footer>


  <!-- SCRIPTS -->
  
  <script  src="https://cdn.jsdelivr.net/npm/nprogress@0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://cdn.jsdelivr.net/npm/nprogress@0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" ></script>
<script  src="https://cdn.jsdelivr.net/npm/bootstrap@4.5.3/dist/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>

<!-- Plugins -->


  
    <script  src="/js/img-lazyload.js" ></script>
  



  



  <script  src="https://cdn.jsdelivr.net/npm/tocbot@4.12.2/dist/tocbot.min.js" ></script>



  <script  src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.5.7/dist/jquery.fancybox.min.js" ></script>



  <script  src="https://cdn.jsdelivr.net/npm/anchor-js@4.3.0/anchor.min.js" ></script>



  <script defer src="https://cdn.jsdelivr.net/npm/clipboard@2.0.8/dist/clipboard.min.js" ></script>




  <script defer src="/js/leancloud.js" ></script>



  <script  src="https://cdn.jsdelivr.net/npm/typed.js@2.0.11/lib/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var title = document.getElementById('subtitle').title;
      
      typing(title)
      
    })(window, document);
  </script>



  <script  src="/js/local-search.js" ></script>
  <script>
    (function () {
      var path = "/local-search.xml";
      $('#local-search-input').on('click', function() {
        searchFunc(path, 'local-search-input', 'local-search-result');
      });
      $('#modalSearch').on('shown.bs.modal', function() {
        $('#local-search-input').focus();
      });
    })()
  </script>





  

  
    <!-- MathJax -->
    <script>
      MathJax = {
        tex: {
          inlineMath: [['$', '$'], ['\\(', '\\)']]
        },
        options: {
          renderActions: {
            findScript: [10, doc => {
              document.querySelectorAll('script[type^="math/tex"]').forEach(node => {
                const display = !!node.type.match(/; *mode=display/);
                const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display);
                const text = document.createTextNode('');
                node.parentNode.replaceChild(text, node);
                math.start = { node: text, delim: '', n: 0 };
                math.end = { node: text, delim: '', n: 0 };
                doc.math.push(math);
              });
            }, '', false],
            insertedScript: [200, () => {
              document.querySelectorAll('mjx-container').forEach(node => {
                let target = node.parentNode;
                if (target.nodeName.toLowerCase() === 'li') {
                  target.parentNode.classList.add('has-jax');
                }
              });
            }, '', false]
          }
        }
      };
    </script>

    <script async src="https://cdn.jsdelivr.net/npm/mathjax@3.1.2/es5/tex-svg.js" ></script>

  








  

  

  

  

  

  





<!-- 主题的启动项 保持在最底部 -->
<script  src="/js/boot.js" ></script>


</body>
</html>
